In [2]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [3]:
import re
import json
from six import iteritems

In [4]:
import sys
sys.path.append('/home/hongfeng/projects/UTIE/')
sys.path.append('/home/hongfeng/projects/math_seq2tree/')

In [8]:
print(sys.path)

['/home/hongfeng/projects/utie_word_problem', '/home/hongfeng/projects/UTIE', '/home/hongfeng/projects/p5', '/home/hongfeng/projects/Bert', '/home/hongfeng/projects/Bert/albert', '/home/hongfeng/projects/math_seq2tree', '/home/hongfeng/miniconda2/envs/torch/lib/python37.zip', '/home/hongfeng/miniconda2/envs/torch/lib/python3.7', '/home/hongfeng/miniconda2/envs/torch/lib/python3.7/lib-dynload', '', '/home/hongfeng/miniconda2/envs/torch/lib/python3.7/site-packages', '/home/hongfeng/miniconda2/envs/torch/lib/python3.7/site-packages/IPython/extensions', '/home/hongfeng/.ipython', '/home/hongfeng/projects/UTIE/Bert', '/home/hongfeng/projects/UTIE/Bert', '/home/hongfeng/projects/UTIE/', '/home/hongfeng/projects/math_seq2tree/']


In [265]:
s = 'a20070522205928AA9RA39'
len(s)

22

In [267]:
s[-21:]
s.lower()

'a20070522205928aa9ra39'

## 若有sources信息，替换全小写的id为sources中正常大小写的id

In [278]:
urls_data = json.load(open('/home/hongfeng/projects/utie_word_problem/data/dolphin18k/eval_urls.json'))

In [ ]:
for word_prob_data in urls_data:
    if "sources" not in word_prob_data:
        continue
    for url in word_prob_data["sources"]:
        if url[-21:].lower() == word_prob_data["id"][-21:]:
            word_prob_data["id"] = "yahoo.answers." + url[-21:]
            print(word_prob_data["id"])

In [280]:
with open('/home/hongfeng/projects/utie_word_problem/data/dolphin18k/eval_urls_new.json', 'w') as f:
    f.write(json.dumps(urls_data, indent=2))

## 统计original_text为空的数目

In [306]:
org_data = json.load(open('/home/hongfeng/projects/utie_word_problem/data/dolphin18k/eval_original.json'))

In [ ]:
count = 0
for d in org_data:
    print(d['original_text'])
    if not d['original_text']:
        count += 1
        print(d)
print(count)

In [329]:
data = json.load(open('/home/hongfeng/projects/utie_word_problem/data/dolphin18k/eval_cleaned.json'))

In [327]:
data = json.load(open('/home/hongfeng/projects/utie_word_problem/data/dolphin18k_data/eval_dataset/eval_dataset_raw.json'))

In [6]:
data = json.load(open('/home/hongfeng/projects/utie_word_problem/data/dolphin18k_data/eval_ids/eval_linear_t6.json'))

In [ ]:
for i in range(10):
    print(data[i])

In [14]:
len(data)

6762

In [149]:
count = 0
for d in data:
    if not d['equations'] or not d['text']:
        count += 1
print(count)

1


In [293]:
count = 0
for d in data:
    if not d['equations']:
        count += 1
print(count)

2373


In [177]:
len([s for s in data[:2000] if s['equations'] and s['text']]), len(data)

(689, 3728)

In [27]:
frac = r'\d+\s?/\s?\d+'
num_re = re.compile(f'\d+ {frac}|{frac}|\d+\.\d+\s?%|\d+\s?%|\d+\s?percent|\d+\.\d+|\d+|(?<![\w)])-\d+\.?\d*')

In [29]:
num_matched = num_re.finditer('1.5%')
for i in num_matched:
    print(i)

<re.Match object; span=(0, 4), match='1.5%'>


In [70]:
import re
s = 'cm'
if re.search('\d+(cm|m)', s):
    print(1)

In [143]:
5.0 == 5.000

True

In [140]:
from word_problem.dolphin.process import parse_answer_equations 

In [141]:
parse_answer_equations('unkn: x\r\nequ: (5/4)/(1/4)=x', None)

(True,
 ['x'],
 [['(', '5', '/', '4', ')', '/', '(', '1', '/', '4', ')', '=', 'x']])

In [647]:
s = 'twenty-eight'
re.search('(?<=[\w)])-', s)

<re.Match object; span=(6, 7), match='-'>

In [ ]:
s1 = ['4', '3', '1', '2']
s2 = {s for s in s1}
s2.index('1')

- not numbers_in_eqs.issubset(numbers_in_text) 855 
- align_nums 56
- successfully matched 197

In [519]:
for i in pairs:
    print(i, pairs[i])
    if len(pairs[i][1]) > 1:
        break

yahoo.answers.20080421132620aaxfxig (['1', 'number', 'is', 'n', 'less', 'than', 'another', '.', 'the', 'difference', 'of', 'n', 'times', 'the', 'smaller', 'and', 'n', 'times', 'the', 'larger', 'number', 'is', 'n', '.', 'what', 'are', 'the', 'numbers', '?'], [['x', '=', 'y', '-', 'N0'], ['N3', '*', 'x', '-', 'N1', '*', 'y', '=', 'N2']], ['4', '5', '-11', '2'], [3, 11, 16, 22], '-3; 1')


In [362]:
pairs['yahoo.answers.20081001213840aa2h0nn']

(['a',
  'number',
  'is',
  'n',
  '+',
  'n',
  'more',
  'than',
  'its',
  'reciprocal',
  '.',
  'find',
  'the',
  'number',
  '.'],
 [['x', '=', 'N2', '/', 'x', '+', 'N1', '+', 'N0']],
 ['1/10', '2', '1'],
 [3, 5])

In [411]:
pairs['yahoo.answers.20080303024828AARMTPE']

(['n', 'of', 'n', 'm', '?'],
 [['N1', '*', 'N0', '=', 'x']],
 ['33', '0.07'],
 [0, 2])

In [586]:
len(pairs)

3730

In [175]:
30/100 == 0.3

True

In [153]:
words, equ, nums, posi = pairs[i]

In [157]:
print(data[2000])

{'original_text': '', 'text': '', 'flag': 0, 'ans': '7; 9', 'equations': 'unkn: x,y\r\nequ: x+2=y\r\nequ: 2*5*x+8*y*2=142', 'id': 'yahoo.answers.20081030152010AAMuQJ1'}


In [158]:
print(words, equ, nums, posi)

['e', '.', 'd', '.', 'c', '.', 'b', '.', 'a', '.'] [] [] []


In [23]:
nums = [['1/3', (0, 3)], ['5', (19, 20)], ['1/2', (31, 34)]]

In [24]:
nums = {n[0]: 'N{}'.format(i) for i, n in enumerate(nums)}

In [13]:
s = '8 percent'
round(float(s[:-7]) / 100., 8)

0.08

In [10]:
1.8 / 100

0.018000000000000002

In [385]:
t = ' is80%and0.5m of what number is 33.6?  '
from word_problem.dolphin.process import *
from word_problem.dolphin.text_num import _extract_nums
print(SpecialNums.handle(t))
print(EnglishToNumber.handle(t))
print(_extract_nums('80% of what number is 33.6 7m 5 / 10 ?'))

80/100 of what number is 33.6?
80% of what number is 33.6 ?
[['80%', (0, 3)], ['33.6', (22, 26)], ['7', (27, 28)], ['5 / 10', (30, 36)]]


In [388]:
t = ' is80%and0.5m of what number is 33.6?  '

In [389]:
segs = t.strip().split()
new_segs = []
for seg in segs:
    nums = _extract_nums(seg)
    if nums:
        pre_end = 0
        for i, n in nums:
            print(i, n)
            if pre_end < n[0]:
                new_segs.append(seg[pre_end:n[0]])
            new_segs.append(seg[n[0]:n[1]])
            pre_end = n[1]
        if pre_end < len(seg):
            new_segs.append(seg[pre_end:])
    else:
        new_segs.append(seg)
print(new_segs)

80% (2, 5)
0.5 (8, 11)
33.6 (0, 4)
['is', '80%', 'and', '0.5', 'm', 'of', 'what', 'number', 'is', '33.6', '?']


In [78]:
numbers, ele_in_text

({'0.08', '1.0', '2.0', '53.0'}, {'1.0', '100.0', '2.0', '53.0', '8.0'})

In [85]:

print(nums)
ele_in_text = set(EqParse.nums_re.findall(new_text))
print(ele_in_text)

[['1', (4, 5)], ['53', (49, 51)], ['00', (52, 54)], ['2', (82, 83)], ['8%', (107, 109)]]
{'8%', '2', '1', '0', '53.0'}


In [27]:
comma_in_num_re = re.compile(r'(\d+),(\d+)(\.\d+)?')

In [33]:
m.groups()

('112', '222', None)

In [74]:
extract_numbers("The one-way fare from Phoenix to Dallas by bus is $53.00. The round-trip fare is double the price less an 8% discount. Find the fare")

('the 1 way fare from phoenix to dallas by bus is $53.00 . the round - trip fare is 2 time the price less an 8% discount . find the fare',
 [['1', (4, 5)],
  ['53', (49, 51)],
  ['00', (52, 54)],
  ['2', (82, 83)],
  ['8%', (107, 109)]])

In [75]:
to_float('8%')

0.08

In [93]:
fraction_re = ('(1|)\s?(half)|(\d+)\s?(third|fou?rth|quarter|fifth|sixth|seventh|eighth|ninth|'
               'tenth|twentieth|thirtieth|fou?rtieth|fiftieth|sixtieth|hundredth)')
re.search(fraction_re, '1quarter').groups()

(None, None, '1', 'quarter')

In [99]:
from usage_example.word_problem.dolphin.text_num import multiple_fraction_to_num
multiple_fraction_to_num('kylie picked 1 seventh of a basket of apples yesterday . she picked 1/2 of a basket of apples today . what part of a basket of apples has kylie picked all together ?')

'kylie picked 1/7 of a basket of apples yesterday . she picked 1/2 of a basket of apples today . what part of a basket of apples has kylie picked all together ?'

In [21]:
from usage_example.word_problem.dolphin.text_num import SpecialNums 
SpecialNums.mixed_number_to_fraction('10 2 / 3')

'32/3'

In [40]:
EqParse.merge_fractions(list('1/3+2/4'), ['1/3', '2/4'])

['1/3', '+', '2/4']

In [17]:
from collections import Counter
Counter([1, 2, 2])

Counter({1: 1, 2: 2})

## 数据预处理

In [ ]:
import urllib
def getHtml(url):
    page = urllib.request.urlopen(url)
    html = page.read()
    return html
print(getHtml("https://answers.yahoo.com/question/index?qid=20080612154712AAvHwGV"))

In [5]:
from word_problem.dolphin.process import infix_to_postfix, Tree, eval_postfix, postfix_to_tree, infix_to_relation_with_equal
from word_problem.math23k.process import add_significant_number
from utie.data_label_to_candidate import reid
from word_problem.math23k.process import to_float
from word_problem.dolphin.process import extract_numbers, EqParse, contain_any, QuestionAnswerParser

In [ ]:
counts = 0
for i in pairs:
    for infix_expression in pairs[i][1]:
#         print(infix_expression)
        print(infix_expression.count('='))

In [434]:
infix_expressions = []
print(infix_expression)
for i, ele in enumerate(infix_expression):
    if ele == '=':
        infix_expressions.append(infix_expression[:i])
        infix_expressions.append(infix_expression[i+1:])
print(infix_expressions)

['N2', '*', 'x', '+', 'N1', '*', 'y', '=', 'N3', '*', 'N0']
[['N2', '*', 'x', '+', 'N1', '*', 'y'], ['N3', '*', 'N0']]


 'relations': [{'id': '[-](N1,1)', 'operands': ('N1', '1'), 'type': '-'}, {'id': '[*](N0,[-](N1,1))', 'operands': ('N0', '[-](N1,1)'), 'type': '*'}, {'id': '[*](N0,N1)', 'operands': ('N0', 'N1'), 'type': '*'}, {'id': '[*](1,N0)', 'operands': ('1', 'N0'), 'type': '*'}, {'id': '[-]([*](N0,N1),[*](1,N0))', 'operands': ('[*](N0,N1)', '[*](1,N0)'), 'type': '-'}]

In [7]:
pairs = {}
count = 0
imp = []
for i, s in enumerate(data):
    if not (s['equations'] and s['text']):
        continue
    try:
        if s['id'] == 'yahoo.answers.20100606182013aaboifr':
            s['equations'] = s['equations'].replace('2*x -11', '2*x - 11')
        if s['id'] == 'yahoo.answers.20080831103207aaxfkfc':
            s['equations'] = s['equations'].replace('3+3/8', '27/8')
        if s['id'] == 'yahoo.answers.20110822104552AA5dVIH':
            s['equations'] = s['equations'].replace('1.4', '(1 + 0.4)')
        if s['id'] == 'yahoo.answers.20110209203154AAE7Sz4':
            s['equations'] = s['equations'].replace('600', '200 * 3')
        if s['id'] == 'yahoo.answers.20070501114420AAd9oYf':
            s['equations'] = s['equations'].replace('/0', '/40')
        if s['id'] == 'yahoo.answers.20071117163907AAzuPdi':
            s['ans'] = '20/3 or 55'
        if s['id'] == 'yahoo.answers.20071117163907AAzuPdi':
            s['ans'] = '1/12'
        if s['id'] == 'yahoo.answers.20080117232720AAfxYrt':
            s['text'] = s['text'].replace('0.3', '. 3')
        if s['id'] == 'yahoo.answers.20080916195022AAQsn1l':
            s['text'] = s['text'].replace('165.000', '165,000')
        qa_p = QuestionAnswerParser(s)
        imp.extend(qa_p.imp)
        if qa_p.pair[1]:
            count += 1
            pairs[s['id']] = qa_p.pair
#             print(qa_p.pair[1])
    except:
        pass

successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
more than one match 2 ['1', '2', '3', '2', '30', '2']
{'type': 0, 'text': 'one number is 2 more than three times a second number. the sum of the two numbers is 30. find the two numbers.', 'sources': ['https://answers.yahoo.com/question/index?qid=20080427194539AAWkEb8'], 'flag': 0, 'original_text': '', 'ans': '{23; 7}', 'equations': 'unkn: x,y\r\nequ: x = 2 + 3*y\r\nequ: x + y = 30', 'id': 'yahoo.answers.20080427194539aawkeb8'}
successfully matched
successfully matched
more than one match 1 ['1', '1']
{'type': 0, 'text': 'What is the sum of 1 and 1?', 'sources': ['https://answers.yahoo.com/question/index?qid=20080505002142AAzgZob'], 'fl

{'type': 0, 'text': 'If x+ 1=3, what is the value of x^4 +1?', 'sources': ['https://answers.yahoo.com/question/index?qid=20080329133816AAGhCRY'], 'flag': 0, 'original_text': '', 'ans': '17', 'equations': 'unkn: x^4 + 1\r\nequ:  x + 1 = 3', 'id': 'yahoo.answers.20080329133816aaghcry'}
successfully matched
successfully matched
successfully matched
{'type': 0, 'text': 'I am a 5-digit number.My tens digit is the sum of my hundreds digit and thousands digit.My tens digit is twice my ones digit.My hundreds digit is 1 more than zero.My thousands digit is seven times my hundreds digit.My ten thousands digit is the same as my ones digit. what is the mystery number?', 'sources': ['https://answers.yahoo.com/question/index?qid=20080615192506AAphMv4', 'https://answers.yahoo.com/question/index?qid=20080701125228AAL7wyV'], 'flag': 0, 'original_text': '', 'ans': '47184', 'equations': 'unkn: m\r\nequ: m = 10000*x + 1000*y + 100*z + 10*a + b\r\nequ: a = z + y\r\nequ: a = 2*b\r\nequ: z = 1 + 0\r\nequ: y 

{1896.0} {1896.0, 41.0, 23.6} {1.0, 2.0, 1896.6, 41.0, 12.0, 23.6}
----------
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
more than one match 2 ['2', '25/100', '2']
{'type': -1, 'text': 'A woman wishes to remodel her kitchen. It is in the shape of a rectangle and she wants to double the area. If she increases the width by 25%, by what percent would she need to increase the length in order to make it double the area?', 'flag': 0, 'original_text': '', 'ans': '0.6', 'equations': 'unkn: x\r\nequ: (1+0.25)(1+x)=2', 'id': 'yahoo.answers.20090312143904aarhkbc'}
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
more than one match 1/4 ['5', '2', '12', '1', '1/2', '1/4', '1/6', '1/3', '1/4', '5']
{'type': -1, 'text': 'Five friends bought a total of 2 pizzas

{3600.0, 5280.0} {3600.0, 2.92, 300.0, 5280.0} {1.0, 2.0, 2.92, 300.0, 60.0}
----------
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
{'type': -1, 'text': 'You drive 45 miles per hour. What is your speed in feet per second?', 'flag': 0, 'original_text': "You drive 45 miles per hour. What is your speed in feet per second?\nI'm in 8th grade, and this is one of my review problems.  I can't see how one could get 66 ft/s.  Any explanations or worked-out solutions would be appreciated.", 'ans': '66', 'equations': 'unkn: x\r\nequ: x = 45*5280/3600', 'id': 'yahoo.answers.20110922170653AA3Mxd0'}
{3600.0, 5280.0} {3600.0, 45.0, 5280.0} {1.0, 2.0, 100.0, 45.0, 60.0}
----------
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
successfully m

{'type': -1, 'text': "If there is a 40% chance of an A for my first subject, 60% for my second, 70% for my third, and 0% for all the rest, what is the probability that I get exactly 2 A's?", 'flag': 0, 'original_text': "What is the probability of this?\nIf there is a 40% chance of an A for my first subject, 60% for my second,  70% for my third, and 0% for all the rest, what is the probability that I get exactly 2 A's?\r\n\r\nThis is a probably I was trying to help someone with, but I couldn't work it out!", 'ans': '0.436', 'equations': 'unkn: x\r\nequ: x=0.400*0.600*0.300+0.400*0.400*0.700+0.600*0.600*0.700', 'id': 'yahoo.answers.20080503004354AA18Sg6'}
{0.3} {0.3, 0.4, 0.6, 0.7} {0.7, 0.4, 0.6, 1.0, 100.0, 0.0, 2.0, 60.0}
----------
successfully matched
successfully matched
{'type': -1, 'text': 'If joe does 5 pages of work each day and after 6 days he he has 60% of this assignment done, how many more pages does he have to do to finish his assignment?', 'flag': 0, 'original_text': 'If 

----------
successfully matched
successfully matched
more than one match 1 ['2643', '447', '1', '1']
{'type': -1, 'text': 'i have 2643 which is my total figure and the figure 447. express this as a ratio 1:X, where 1 is the total number. Find the value of x.', 'flag': 0, 'original_text': "ratio help please?\ni am total rubbish with ratio's so could someone please help me express the following\r\n\r\ni have 2643 which is my total figure and the figure 447.  i need to express this as a ratio 1:X.  where 1 is the total number?\r\n\r\nany help even explanations of how to do it would be greatly appreciated", 'ans': '0.169', 'equations': 'unkn: x\r\nequ: 1/x=2643/447', 'id': 'yahoo.answers.20080509105159AAfAtU0'}
successfully matched
{'type': -1, 'text': 'in a flock of black and white sheep, 2 out of 5 sheep are white.  if there are 8 more black sheep than white sheep, how many sheep are in the flock?', 'flag': 0, 'original_text': 'problem solving?\nin a flock of black and white sheep, 2 out

more than one match 1 ['1', '6', '1', '4', '15', '2.4', '24', '36', '15']
{'type': -1, 'text': 'Ray can pick all of the oranges in 1 row of a frove in 6 hrs. Asa can pick all of the orange in 1 row of the gorve in 4 hrs. If they work together, how long would it take them to pick 15 rows of the grove? A. 2.4 hr\nB. 24 hr\nC. 36 hr\nD. 15 hr', 'flag': 0, 'original_text': 'Please explain this math problem!?\nRay can pick all of the oranges in 1 row of a frove in 6 hrs. Asa can pick all of the orange in 1 row of the gorve in 4 hrs. If they work together, how long would it take them to pick 15 rows of the grove?\r\n\r\nA. 2.4 hr\r\nB. 24 hr\r\nC. 36 hr\r\nD. 15 hr\r\n\r\nPlease show your work.', 'ans': '36', 'equations': 'unkn: x\r\nequ: (1/6+1/4)*x=15', 'id': 'yahoo.answers.20070502171730AAjCl56'}
successfully matched
successfully matched
successfully matched
more than one match 1 ['2', '4', '6', '26', '2', '1', '1', '1']
{'type': -1, 'text': 'There are 2 large blocks and 4 small blocks on

more than one match 3 ['7', '4', '3', '4', '3']
{'type': -1, 'text': '7 disks in a bag (4 red & 3 blk).  To win, you must draw the 4 red disks out of the bag before you draw the 3 blk disks.  &quot;three strikes and youre out&quot;.  What is the theoretical probability of winning and losing?  If that cant be calculated, then why?', 'flag': 0, 'original_text': 'Probability  question?.\n7 disks in a bag (4 red & 3 blk).  To win, you must draw the 4 red disks out of the bag before you draw the 3 blk disks.  &quot;three strikes and youre out&quot;.  What is the theoretical probability of winning and losing?  If that cant be calculated, then why?.\n\n\n', 'ans': '0.571428571429; 0.428571428571', 'equations': 'unkn: x,y\r\nequ: x=4/7\r\nequ: y=3/7', 'id': 'yahoo.answers.20100330231612AAKhjJC'}
successfully matched
more than one match 2 ['2', '1', '3/4', '1', '5/3', '2', '4']
{'type': -1, 'text': 'Two employees are painting a boat. one uses 3/4 gallon of paint each hour, one uses 1 2/3 gallon

more than one match 4 ['-6', '-1', '4', '9', '14', '-6', '-1', '4', '9', '14']
{'type': -1, 'text': 'Find the common difference between the terms of the following arithmetic sequence.  -6, -1, 4, 9, 14, .? Find the common difference between the terms of the following arithmetic sequence. \n-6, -1, 4, 9, 14, ..', 'flag': 0, 'original_text': 'Find the common difference between the terms of the following arithmetic sequence.  -6, -1, 4, 9, 14, ..?\nFind the common difference between the terms of the following arithmetic sequence.\r\n\r\n-6, -1, 4, 9, 14, ..', 'ans': '5', 'equations': 'unkn: x\r\nequ: 4+1=x', 'id': 'yahoo.answers.20090407195326AAIzScL'}
successfully matched
{'type': -1, 'text': 'To play this game four face cards and two aces are taken out of an ordinary deck of cards. You shuffle them and then the dealer lays the cards out in a row, face down. The dealer bets you $5 that you cannot pick two cards out of the six without picking an ace. If you do, you win $5. If you pick one

{0.6} {0.6} {0.0, 1.0, 2.0, 4.0, 9.0}
----------
successfully matched
{'type': -1, 'text': 'sqrt 98 minus sqrt 128 divided by sqrt 8 minus sqrt 2.', 'flag': 0, 'original_text': 'Use the power rule?\nHi.\r\nI have the following and want to use the power rule to solve it. How do I do it?\r\nsqrt 98 minus sqrt 128 divided by sqrt 8 minus sqrt 2.\r\nThank you', 'ans': '-1', 'equations': 'unkn: o\r\nequ: o=(-1)', 'id': 'yahoo.answers.20070331101314AADr9lj'}
{-1.0} {-1.0} {128.0, 1.0, 98.0, 2.0, 100.0, 8.0, 60.0}
----------
successfully matched
{'type': -1, 'text': 'A triangle with sides of 5, 12, 13 cm is similar to a triangle whos longest side is 39cm. What is the perimeter of the largest triangle? Also what is the ratio of the perimeter and area of the small triangle to the large one?', 'flag': 0, 'original_text': 'Some questions on geometry?\n\n\n\nA triangle with sides of 5,12,13 cm is similar to a triangle whos longest side is 39cm.\n\n\n\nWhat is the perimeter of the largest triangle?

{18.0, 36.0} {18.0, 36.0} {1.0, 2.0}
----------
{'type': -1, 'text': 'What is 3 3/5 divided by 4 1/2 ?', 'flag': 0, 'original_text': 'What is 3 3/5 divided by 4 1/2 ?\n', 'ans': '0.8', 'equations': 'unkn: x\r\nequ: 18/5*2/9=x', 'id': 'yahoo.answers.20100215114504AAY2avw'}
{9.0} {9.0, 2.0, 3.6} {1.0, 2.0, 3.6, 4.5}
----------
more than one match 6 ['6', '255', '15', '6']
{'type': -1, 'text': 'I multiply it by 6 and add 255.. I get the same answer if i multiply by 15 and subtract 6. \nwhat is my number ?', 'flag': 0, 'original_text': 'I am thinking of a number ...?.\nI am thinking of a number..\nI multiply it by 6 and add 255..\nI get the same answer if i multiply by 15 and subtract 6..\nwhat is my number ??.\nPlease (ANYONE) answer my question .\nThanks :).\n\n\n', 'ans': '29', 'equations': 'unkn: x\r\nequ: x*6+255=x*15-6', 'id': 'yahoo.answers.20100219073208AAgywyB'}
successfully matched
successfully matched
{'type': -1, 'text': 'an aircraft flies 10km south and then 6km west. what is 

----------
successfully matched
successfully matched
successfully matched
more than one match 8 ['8', '32/5', '8']
{'type': -1, 'text': 'A rectangle had an area of 8cm. If the length is 6 2/5 (Six and Two Fiths) what is the width?', 'flag': 0, 'original_text': 'Maths Help Please!   FRACTIONS!?\n\n\n\nI have an exam on Tuesday and need someone to give me the answer to this question with an explanation please.\n\n\n\nThank You x\n\n\n\nA rectangle had an area of 8cm2. If the length is 6 2/5 (Six and Two Fiths) what is the width?\n\n\n\n.........................................\n\n', 'ans': '1.25', 'equations': 'unkn: x\r\nequ: 8=6.4*x', 'id': 'yahoo.answers.20090503075542AAlkgNY'}
successfully matched
successfully matched
successfully matched
successfully matched
{'type': -1, 'text': 'A package contains 9 candy canes, 6 of which are cracked. If 2 are selected, find the probability of getting no cracked candy canes.', 'flag': 0, 'original_text': 'A package contains 9 candy canes, 6 of whi

{'type': -1, 'text': 'If you draw one card at random from a deck of 15 cards numbered 1 through 15,? inclusive, what is the probability that the number you draw is divisible by 3 and odd?', 'flag': 0, 'original_text': 'If you draw one card at random from a deck of 15 cards numbered 1 through 15,?.\ninclusive, what is the probability that the number you draw is divisible by 3 and odd?.\n\n\n', 'ans': '0.2', 'equations': 'unkn: y\r\nequ: y=3/15', 'id': 'yahoo.answers.20090207230351AA9Q2XQ'}
{'type': -1, 'text': 'How do you find 1/20th of 1/100th of 50?', 'flag': 0, 'original_text': 'How do you find 1/20th of 1/100th of 50?\nI need help and show me your work of how you did it!', 'ans': '0.025', 'equations': 'unkn: x\r\nequ: 0.0005*50=x', 'id': 'yahoo.answers.20090208142859AAPfnGh'}
{0.0005} {50.0, 0.0005} {0.05, 1.0, 2.0, 0.01, 50.0}
----------
{'type': -1, 'text': 'during a certain period of days, observations were made on the weather. It was noted that whenever it rained in the morning 

{6.5} {1.0, 2.0, 6.5} {1.0, 2.0, 100.0, 6.0, 60.0, 30.0}
----------
{'type': -1, 'text': 'If i make 9 dollars an hour and work full time 8 hours a day, 5 days a week, what is my anual income?', 'flag': 0, 'original_text': 'If i make 9 dollars an hour and work full time 8 hours a day, 5 days a week, what is my anual income?\n\n', 'ans': '72; 360; 18720', 'equations': 'unkn: x,y,z\r\nequ: 9*8=x\r\nequ: x*5=y\r\nequ: y*52=z', 'id': 'yahoo.answers.20060622090344AAtpd5Q'}
{52.0} {8.0, 9.0, 52.0, 5.0} {1.0, 2.0, 5.0, 8.0, 9.0, 60.0}
----------
successfully matched
{'type': -1, 'text': "how many ways can you arrange numbers 12345678 in groups of 4's?", 'flag': 0, 'original_text': "how many ways can you arrange numbers 12345678in groups of 4's? Please show work?\nexamples 1234\r\n              5678\r\n              2436", 'ans': '1680', 'equations': 'unkn: x\r\nequ: 8*7*6*5=x', 'id': 'yahoo.answers.20060717103530AANVNdk'}
{8.0, 5.0, 6.0, 7.0} {8.0, 5.0, 6.0, 7.0} {1.0, 2.0, 4.0, 12345678.0}
--

{0.023} {0.023, 10.0} {1.0, 2.3, 2.0, 100.0, 10.0, 60.0}
----------
more than one match 4 ['4', '1', '2', '3', '4', '2', '5']
{'type': -1, 'text': 'A box contains 4 slips of paper, on each of which is written the number 1, 2, 3, or 4, respectively.  A slip is drawn and the number on it is noted.  That slip is put aside, another is drawn, and the number on it is noted. What is the probability that the sum of the two numbers is 5?', 'flag': 0, 'original_text': 'I REALLY NEED HELP WITH THIS PROBABILITY QUESTION!!!!!!?\n\n\n\nCan you please explain the problem to me.  I now the answer is 1/3 but I want an EXPlanation\n\nThank you!\n\n:\n\nA box contains 4 slips of paper, on each of which is written the number 1, 2, 3, or 4, respectively.  A slip is drawn and the number on it is noted.  That slip is put aside, another is drawn, and the number on it is noted. What is the probability that the sum of the two numbers is 5?\n\n', 'ans': '12; 0.333333333333', 'equations': 'unkn: x,z\r\nequ: 4*3=x

successfully matched
more than one match 15 ['100', '20', '3', '15', '15']
{'type': -1, 'text': 'Karen is riding in a 100-mile bike race.  She pedals 20 mi per hour for 3 hours.  She pedals 15 mi per hour for the rest of the race.  How long does Karen pedal 15 miles an hour?', 'flag': 0, 'original_text': 'Karen is riding in a 100-mile bike race.  She pedals 20 mi per hour for 3 hours.  She pedals 15 mi per hour..?\n\n\n\n... for the rest of the race.  How long does Karen pedal 15 miles an hour?\n\n', 'ans': '2.66666666667', 'equations': 'unkn: x\r\nequ: 20*3+15*x=100', 'id': 'yahoo.answers.20080424180846AAU5MJY'}
more than one match 5 ['5', '8', '3', '5', '2']
{'type': -1, 'text': 'Look at these expressions. 5y - 8 - fist expression\n3y + 5 - second expression. what value of y makes the two expressions equal ?', 'flag': 0, 'original_text': 'maths equation ?!?!?\n\n\n\nLook at these expressions\n\n\n\n5y - 8 - fist expression\n\n3y + 5 - second expression\n\n\n\nwhat value of y makes th

{2760.0, 12.0, 7.0, 2300.0} {2760.0, 12.0, 2300.0, 7.0} {0.1, 0.12, 0.07, 1.0, 100.0, 2.0, 230.0, 1000.0, 60.0}
----------
successfully matched
{'type': -1, 'text': 'if drawing 2 cards from a deck, find the probability that the 2nd card is the 4 of hearts if 1st cardis black? a. 1/52\nb. 1/51\nc. 4/51\nd. 1/2652\ne. none of these', 'flag': 0, 'original_text': 'if drawing 2 cards from a deck, find the probability that the 2nd card is the 4 of hearts if 1st cardis black?\na. 1/52\r\nb. 1/51\r\nc. 4/51\r\nd. 1/2652\r\ne. none of these', 'ans': '0.0192307692308; 0.0098', 'equations': 'unkn: x,y\r\nequ: 51/52*1/51=x\r\nequ: 26/52*1/51=y', 'id': 'yahoo.answers.20080310211032AAzwu3x'}
{26.0, 51.0, 52.0} {0.01961, 26.0, 51.0, 52.0} {0.07843, 1.0, 2.0, 4.0, 0.01923, 0.00038, 0.01961}
----------
successfully matched
successfully matched
successfully matched
successfully matched
successfully matched
{'type': -1, 'text': 'If  39% of someones income is 44,850, how much is the the rest of the 61%?',

successfully matched
{'type': -1, 'text': 'Farmer had 100 cows in winter. In spring the herd increased by 10%. In summer farmer sold 10% of animals. \nHow many cows left?', 'flag': 0, 'original_text': 'Farmer had 100 cows in winter. In spring the herd increased by 10%. In summer farmer sold 10% of animals.?\nHow many cows left?', 'ans': '99', 'equations': 'unkn: x\r\nequ: 110-11=x', 'id': 'yahoo.answers.20071030145159AAZTeTP'}
{11.0, 110.0} {11.0, 110.0} {0.1, 1.0, 2.0, 100.0}
----------
successfully matched
{'type': -1, 'text': 'What is the LCD (least common multiple) of 6, 7, 8?', 'flag': 0, 'original_text': 'What is the LCD (least common multiple) of 6, 7, 8?\n\n', 'ans': '168', 'equations': 'unkn: x\r\nequ: 2*3*7*4=x', 'id': 'yahoo.answers.20071106171237AARQNSZ'}
{3.0, 4.0} {2.0, 3.0, 4.0, 7.0} {1.0, 2.0, 6.0, 7.0, 8.0}
----------
successfully matched
{'type': -1, 'text': 'This is what is says on the paper\nInteger Subtraction; range (-99) to (+99) (G)\nQuestion:   (+9) - (-69)', '

----------
{'type': -1, 'text': 'A shop has this special offer reduction of 5% when your bill is between 30 pound and 60 pound, reduction of 10% when your bill is more than 60 pound. before the reductions, mels bill is 48 pound and bills bill is 54 pound. After the reductions, who paid more?', 'flag': 0, 'original_text': 'percentage maths question? help please?.\na shop has this special offer..\n&quot;reduction of 5% when your bill is between 30pound and 60pound..\nreduction of 10% when your bill is more than 60pound.&quot;.\nbefore the reductions, mels bill is 48pound and bills bill is 54pound..\nafter the reductions, who paid more?.\nyou must show working to explain your answer..\ntick the correct answer: mel, both paid the same, bill..\ncan someone pleaseeeeeeeee help me??.\n\n\n', 'ans': '45.6; 51.3', 'equations': 'unkn: x,y\r\nequ: 0.95*48=x\r\nequ: 0.95*54=y', 'id': 'yahoo.answers.20100325123424AA4yQlH'}
{0.95} {48.0, 0.95, 54.0} {0.1, 0.05, 1.0, 2.0, 100.0, 48.0, 54.0, 60.0, 30.

{4000000000.0} {4000000000.0, 1800000000.0} {1800000000.0, 1.0, 2.0, 40000000.0, 100.0}
----------
successfully matched
{'type': -1, 'text': 'It has to do with the input and output problems. the table says. Input: 7, 1, 2. While the Output: 60, 24, 30, 36. we need to find the rule in order to find the input for 36.', 'flag': 0, 'original_text': 'Can I have help with this math question?.\nmy little sister who is 9 years old is having trouble with her homework and i forgot how to do this. It has to do with the input and output problems. the table says... Input:7, 1, 2. While the Output: 60, 24, 30, 36. we need to find the rule in order to find the input for 36. thank you!.\n\n\n', 'ans': '3', 'equations': 'unkn: z\r\nequ: 36=6*z+18', 'id': 'yahoo.answers.20090512175513AAMqWFw'}
{18.0, 6.0} {18.0, 36.0, 6.0} {1.0, 2.0, 36.0, 7.0, 24.0, 60.0, 30.0}
----------
successfully matched
successfully matched
successfully matched
successfully matched
{'type': -1, 'text': 'what is the smallest 4 dig

----------
more than one match 2.3 ['2.3', '2.3', '0.3']
{'type': -1, 'text': 'We have built our kids a sandpit and need to know how much sand to put in it.The sandpit dimentions are:\n2.3m long x 2.3 m wide x 0.3 metres deep.', 'flag': 0, 'original_text': 'Stupid Math Question about volume/mass?\n\n\n\nWe have built our kids a sandpit and need to know how much sand to put in it.asically I have forgotten how to work out the volume of a shape. The sandpit dimentions are:\n\n2.3m long x 2.3 m wide x .3 metres deep.\n\nCan you refresshmy brain on this so simple maths problems.\n\nThanks\n\n', 'ans': '1.587', 'equations': 'unkn: x\r\nequ: 2.3*2.3*0.3=x', 'id': 'yahoo.answers.20071126010956AAkHTty'}
{'type': -1, 'text': 'There are 71 stamps cost 19 pence by rounding find the approximate cost of 71 stamps', 'flag': 0, 'original_text': 'a maths problem?\n\n\n\nThere are 71 stamps cost 19 pence by rounding find the approxcost of 71 stamps\n\n', 'ans': '1400', 'equations': 'unkn: x\r\nequ: 20*7

{289.0, 350.0} {289.0, 350.0} {64.0, 1.0, 2.0, 4.0, 5.0, 70.0, 71.0, 75.0, 79.0}
----------
successfully matched
{'type': -1, 'text': 'A bag contains 6 chipps, numbered 1 thru 6. if 2 chips are chosen at random without replacement and the vallues on those 2 chips are multiplied, what is the probability that this product will be greater than 20?', 'flag': 0, 'original_text': 'How do I figure out this SAT math question?\n\n\n\na bag contains 6 chipps, numbered 1 thru 6. if 2 chips are chosen at random without replacement and the vallues on those 2 chips are multiplied, what is the probability that this product will be greater than 20??\n\n', 'ans': '30; 0.133333333333', 'equations': 'unkn: x,y\r\nequ: 6*5=x\r\nequ: 4/x=y', 'id': 'yahoo.answers.20080912205810AAAp3WQ'}
{4.0, 5.0} {4.0, 5.0, 6.0} {1.0, 2.0, 20.0, 6.0}
----------
successfully matched
successfully matched
more than one match 2 ['3', '2', '2']
{'type': -1, 'text': '3 girls & 2 boys sit in a row. How many different ways can the

{16.0, 4.0} {16.0, 4.0} {1.0, 10.0, 2.0, 10100.0}
----------
successfully matched
successfully matched
successfully matched
{'type': -1, 'text': 'A cashier has a total of 124 bills, made up of fives and tens. The total value of the money is $660. How many ten-dollar bills does the cashier have?', 'flag': 0, 'original_text': 'A cashier has a total of 124 bills?\nA cashier has a total of 124 bills, made up of fives and tens.  The total value of the money is $660.  How many ten-dollar bills does the cashier have?', 'ans': '116; 8', 'equations': 'unkn: x,y\r\nequ: 124=x+y\r\nequ: 660=5*x+10*y', 'id': 'yahoo.answers.20080125112426AA16z2e'}
{5.0} {124.0, 10.0, 660.0, 5.0} {1.0, 2.0, 10.0, 660.0, 124.0}
----------
more than one match 3 ['3', '4', '5', '3']
{'type': -1, 'text': 'Angel goes swimming every 3rd day. \nPaul goes swimming every 4th day. \nEmily goes swimming every 5th day. \nToday they were all at the pool together\n\nIn how many days will the 3 friends meet at the pool again?', 'f

In [8]:
count

4243

In [31]:
pairs['yahoo.answers.20090510192349AAqXyhe']

(['n',
  'of',
  'the',
  'ninth',
  'grade',
  'students',
  'at',
  'richbartville',
  'hs',
  'take',
  'algebra',
  '.',
  'if',
  'n',
  'ninth',
  'grade',
  'students',
  'take',
  'algebra',
  ',',
  'how',
  'many',
  'ninth',
  'grade',
  'students',
  "don't",
  'take',
  'algebra',
  '.'],
 [['N1', '/', 'N0', '=', 'x', '/', '(', '1', '-', 'N0', ')']],
 ['90/100', '180'],
 [0, 13],
 {'1', '100', '2'},
 '20')

### 数据总个数 6762，6761个有效，其中
- 有些数字公式中有，文本中没有的sample数 1614
- more than one match 722
- successfully matched 4406
- 公式中包含特殊运算符 is_digital()等 20个

### 统计公式中有，文本中没有的单词

In [565]:
from collections import Counter
Counter(imp).most_common()

[(4.0, 66),
 (3.0, 65),
 (5.0, 57),
 (6.0, 38),
 (9.0, 31),
 (8.0, 27),
 (10.0, 27),
 (16.0, 19),
 (7.0, 19),
 (12.0, 17),
 (20.0, 16),
 (24.0, 15),
 (15.0, 15),
 (0.5, 12),
 (0.75, 11),
 (25.0, 11),
 (180.0, 10),
 (0.8, 10),
 (14.0, 9),
 (13.0, 9),
 (360.0, 9),
 (11.0, 8),
 (90.0, 8),
 (18.0, 8),
 (36.0, 8),
 (60.0, 8),
 (0.4, 7),
 (-1.0, 7),
 (100.0, 7),
 (1.5, 7),
 (40.0, 7),
 (0.0, 7),
 (5280.0, 6),
 (2.5, 6),
 (0.85, 6),
 (52.0, 6),
 (17.0, 5),
 (1000.0, 5),
 (26.0, 5),
 (150.0, 5),
 (0.15, 5),
 (50.0, 5),
 (48.0, 5),
 (231.0, 5),
 (-3.0, 4),
 (0.25, 4),
 (30.0, 4),
 (0.06, 4),
 (120.0, 4),
 (64.0, 4),
 (2.54, 4),
 (144.0, 4),
 (365.0, 4),
 (19.0, 4),
 (28.0, 4),
 (4.5, 4),
 (3.5, 4),
 (0.1, 3),
 (3600.0, 3),
 (100000.0, 3),
 (65.0, 3),
 (33.0, 3),
 (32.0, 3),
 (55.0, 3),
 (0.7, 3),
 (84.0, 3),
 (0.6, 3),
 (34.0, 3),
 (0.38, 3),
 (130.0, 3),
 (27.0, 3),
 (200.0, 3),
 (1024.0, 2),
 (39.37, 2),
 (29.0, 2),
 (1.609, 2),
 (1.8, 2),
 (0.3, 2),
 (43560.0, 2),
 (1.28, 2),
 (0.95, 2),
 (1

### 答案校正

In [575]:
from sympy import *
import regex as re
x = Symbol('x')
y = Symbol('y')
z = Symbol('z')
unks = [x, y, z]
b = Symbol('b')
cnt = 0
ans_dict = {}
for idx in pairs:
    p = pairs[idx]
    eqs = p[1]
    values = p[2]
    new_eqs = []
    r_ans = p[-1]
    for eq in eqs:
        new_eq = []
        for e in eq:
            if e[0] == 'N':
                e = values[int(e[1:])]
                if '/' in e:
                    new_eq.append('({})'.format(e))
                else:
                    new_eq.append(e)
            else:
                new_eq.append(e)
        new_eq = ''.join(new_eq).split('=')
        new_eq = '{}-({})'.format(new_eq[0], new_eq[1])
        try:
            new_eqs.append(eval(new_eq))
        except:
            print('wrong format of new_eq: ', new_eq, p, data_dict[idx])
    if '|' in r_ans:
        r_ans = r_ans.split('|')[0]
    if 'or' in r_ans:
        r_ans = r_ans.split('or')
        
    try:
        if isinstance(r_ans, list):
            new_r_ans = []
            for ans in r_ans:
                ans = ans.strip('{} ?$').split(';')
                if isinstance(ans, str):
                    ans = [ans]
                for i, a in enumerate(ans):
                    new_r_ans.append((unks[i], round(to_float(re.sub('km\^2|/?[a-zA-Z]|\s', '', a)), 3)))
            r_ans = set(new_r_ans)
        else:
            r_ans = r_ans.strip('{} ?$').split(';')
            r_ans = {(unks[i], round(to_float(re.sub('km\^2|/?[a-zA-Z]|\s', '', a)), 3)) for i, a in enumerate(r_ans)}
    except:
        print('wrong format of real answer:', r_ans,  p, data_dict[idx])
    p_ans = solve(new_eqs)
    try:
        if isinstance(p_ans, list):
            new_p_ans = []
            for ans in p_ans:
                for k, v in ans.items():
                    new_p_ans.append((k, round(float(v), 3)))
            p_ans = set(new_p_ans)
        else:
            p_ans = {(k, round(float(v), 3)) for k, v in p_ans.items()}
    except:
        print('wrong format of predict answer: ', type(p_ans), p_ans,  p, data_dict[idx])
    if isinstance(p_ans, set) and isinstance(r_ans, set) and r_ans.issubset(p_ans):
        cnt += 1
    else:
        print(new_eqs)
        print('r_ans, p_ans: ', r_ans, p_ans, p, data_dict[idx])
    ans_dict[idx] = p_ans

[x + y - 825, x - 8*y - 15]
r_ans, p_ans:  {(x, 725.0), (y, 90.0)} {(x, 735.0), (y, 90.0)} (['the', 'sum', 'of', '2', 'numbers', 'is', 'n', '.', 'dividing', 'the', 'larger', 'number', 'by', 'the', 'smaller', 'number', 'yields', 'n', 'with', 'a', 'remainder', 'of', 'n', '.', 'what', 'are', 'the', '2', 'numbers', '?'], [['x', '+', 'y', '=', 'N0'], ['x', '=', 'N1', '*', 'y', '+', 'N2']], ['825', '8', '15'], [6, 17, 22], {'1', '2'}, '{725; 90}') {'type': 0, 'text': 'the sum of two numbers is 825.  Dividing the larger number by the smaller number yields 8 with a remainder of 15. What are the 2 numbers?', 'sources': ['https://answers.yahoo.com/question/index?qid=20100103174417AAtRonT'], 'flag': 0, 'original_text': '', 'ans': '{725; 90}', 'equations': 'unkn: m,n\r\nequ: m+n=825\r\nequ: m = 8*n + 15', 'id': 'yahoo.answers.20100103174417aatront'}
[x - 2*y, x - y - 32]
r_ans, p_ans:  {(y, -32.0), (x, -64.0)} {(y, 32.0), (x, 64.0)} (['1', 'number', 'is', 'n', 'times', 'another', 'number', '.', 't

[x*(x + 1) + 2*x - 270]
r_ans, p_ans:  {(x, -18.0), (x, 15.0), (y, 16.0), (y, -17.0)} {(x, -18.0), (x, 15.0)} (['the', 'sum', 'of', '2', 'consecutive', 'integers', 'and', 'their', 'product', 'is', 'n', '.', 'find', 'the', 'integers', '?'], [['x', '+', '(', 'x', '+', '1', ')', '+', 'x', '*', '(', 'x', '+', '1', ')', '=', 'N0']], ['271'], [10], {'1', '2'}, '15; 16 or -18; -17') {'type': 0, 'text': 'the sum of 2 consecutive integers and their product is 271. find the integers?', 'sources': ['https://answers.yahoo.com/question/index?qid=20090218015526AAmWSDN'], 'flag': 0, 'original_text': '', 'ans': '15; 16 or -18; -17', 'equations': 'unkn: n,n+1\r\nequ: n + (n+1) + n(n+1) = 271', 'id': 'yahoo.answers.20090218015526aamwsdn'}
[1.54285714285714*z - 58.6285714285714]
r_ans, p_ans:  {(y, 77.0), (x, 75.0)} {(z, 38.0)} (['find', 'n', 'consecutive', 'odd', 'numbers', 'such', 'that', 'the', 'sum', 'of', 'n', 'of', 'the', 'smaller', 'and', 'n', 'of', 'the', 'larger', 'is', 'equal', 'to', 'n', '.'],

[x - y - 3, -6*y + z + 1, -x - y + z - 24]
r_ans, p_ans:  {(x, 41.0)} {(z, 41.0), (x, 10.0), (y, 7.0)} (['a', 'man', 'has', 'a', 'daughter', 'and', 'a', 'son', '.', 'the', 'son', 'is', 'n', 'years', 'older', 'than', 'the', 'daughter', 'is', 'now', '.', 'in', 'n', 'year', ',', 'the', 'man', 'will', 'be', 'n', 'times', 'as', 'old', 'as', 'the', 'daughter', 'is', 'now', 'and', 'in', 'n', 'years', ',', 'he', 'will', 'be', 'n', 'years', 'older', 'than', 'the', 'combined', 'ages', 'of', 'his', 'children', '.', 'what', 'is', 'his', 'present', 'age', '?'], [['x', '-', 'N0', '=', 'y'], ['z', '=', 'N2', '*', 'y', '-', 'N1'], ['z', '+', 'N3', '=', 'x', '+', 'N3', '+', 'y', '+', 'N3', '+', 'N4']], ['3', '1', '6', '10', '14'], [12, 22, 29, 40, 46], {'1', '12', '2'}, '41') {'text': 'A man has a daughter and a son. The son is 3 years older than the daughter is now. in one year, the man will be 6 times as old as the daughter is now and in 10 years, he will be 14 years older than the combined ages of h

[x - 0.04, y - 1/x]
r_ans, p_ans:  {(x, 25.0)} {(x, 0.04), (y, 25.0)} (['f', '(', 'x', ')', '=', 'n', '/x', '.', 'g', '(', 'x', ')', '=', 'x^', '2', '.', 'what', 'is', 'f', '(', 'g', '(', 'n', ')', ')', '?'], [['x', '=', 'N1', '*', 'N1'], ['y', '=', 'N0', '/', 'x']], ['1', '0.2'], [5, 22], {'1', '2'}, '25') {'type': -1, 'text': 'f(x)=1/x. g(x)=x^2. What is f(g(0.2))?', 'flag': 0, 'original_text': '', 'ans': '25', 'equations': 'unkn: g\r\nequ: g=0.2*0.2\r\nequ: x = 1/g', 'id': 'yahoo.answers.20080915155120aaaqnha'}
[x - 0.5*y - 5, 2*x + 2*y - 70]
r_ans, p_ans:  {(x, 20.0), (y, 15.0)} {(y, 20.0), (x, 15.0)} (['if', 'the', 'width', 'if', 'a', 'rectangle', 'is', 'n', 'inches', 'more', 'than', 'n', 'of', 'its', 'length', 'and', 'the', 'perimeter', 'of', 'the', 'rectangle', 'is', 'n', 'inches', 'what', 'are', 'the', 'length', 'and', 'width'], [['x', '=', 'N1', '*', 'y', '+', 'N0'], ['(', 'y', '+', 'x', ')', '*', '2', '=', 'N2']], ['5', '1/2', '70'], [7, 11, 22], {'1', '12', '2'}, '20; 15') {

[139*x - 52600.0]
r_ans, p_ans:  {(x, 378.42)} {(x, 378.417)} (['in', 'a', 'group', 'of', 'n', 'acura', 'integras', ',', 'n', 'had', 'a', 'defective', 'alternator', '.', 'if', 'there', 'were', 'n', 'acura', 'integras', 'produced', ',', 'how', 'many', 'would', 'you', 'expect', 'to', 'have', 'defective', 'timing', 'belts', '?'], [['N0', '*', 'x', '=', 'N2', '/', 'N1']], ['139', '1', '52600'], [4, 8, 17], {'1', '2'}, '378.42') {'type': -1, 'text': 'In a group of 139 Acura Integras, one had a defective alternator. If there were 52,600 Acura Integras produced, how many would you expect to have defective timing belts?', 'flag': 0, 'original_text': '', 'ans': '378.42', 'equations': 'unkn: t\r\nequ: 139t=52600/1', 'id': 'yahoo.answers.20090408145009aa0zve0'}
[x + y - 9, -3*x + 3*y - 9]
r_ans, p_ans:  {(x, 6.0)} {(y, 6.0), (x, 3.0)} (['a', 'penguin', 'swimming', 'with', 'the', 'ocean', 'current', 'in', 'search', 'of', 'food', 'takes', 'n', 'hour', 'to', 'travel', 'n', 'miles', '.', 'its', 'retu

[3*x - 201]
r_ans, p_ans:  {(y, 68.0), (z, 69.0), (x, 67.0)} {(x, 67.0)} (['3', 'consecutive', 'numbers', 'add', 'to', 'n', ',', 'what', 'are', 'these', 'numbers', '?'], [['x', '+', '(', 'x', '+', '1', ')', '+', '(', 'x', '+', '2', ')', '=', 'N0']], ['204'], [5], {'1', '2'}, '67; 68; 69') {'type': -1, 'text': 'Three consecutive numbers add to 204, what are these numbers?', 'flag': 0, 'original_text': '', 'ans': '67; 68; 69', 'equations': 'unkn: t\r\nequ: t+(t+1)+(t+2)=204', 'id': 'yahoo.answers.20100219193338aa7vntp'}
[0.7*x - 75.69]
r_ans, p_ans:  {(x, 108.13)} {(x, 108.129)} (['amy', 'paid', '$', 'n', 'for', 'a', 'pair', 'running', 'shoes', 'during', 'a', 'n', 'off', 'sale', '.', 'what', 'was', 'the', 'regular', 'price', '?', 'round', 'to', 'the', 'nearest', 'cent', 'if', 'necessary'], [['x', '*', '(', '1', '-', 'N1', ')', '=', 'N0']], ['75.69', '30/100'], [3, 11], {'1', '100', '2'}, '108.13') {'type': -1, 'text': 'amy paid $75.69 for a pair running shoes during a 30% off sale. what 

[x + y - 357, -7*x + 10*y]
r_ans, p_ans:  {(y, 210.0), (x, 147.0)} {(y, 147.0), (x, 210.0)} (['there', 'are', 'n', 'boys', 'and', 'girls', 'in', 'a', 'class', 'the', 'ratio', 'is', 'n', 'to', 'n', '.', 'how', 'many', 'boys', 'are', 'in', 'the', 'class', '.'], [['y', '+', 'x', '=', 'N0'], ['N2', '*', 'y', '=', 'N1', '*', 'x']], ['357', '7', '10'], [2, 12, 14], {'1', '2'}, '147; 210') {'type': -1, 'text': 'there are 357 boys and girls in a class  the ratio is 7 to 10. how many boys are in the class.', 'flag': 0, 'original_text': 'Ratio Question?\ni knwo this is an easy question but i havent done ratio since the ninth grade and i have to take a state exam tommorow and it will be on there. anyways cna someone answer this and tell me how u did it.  there are 357 boys and girls in a class  the ratio is 7 to 10 how many boys are in the class . thank u in advance', 'ans': '147; 210', 'equations': 'unkn: n\r\nequ: m + n = 357\r\nequ: 10m = 7n', 'id': 'yahoo.answers.20080123132056AAQXKrJ'}
[x - 

[x - 94.2]
r_ans, p_ans:  {(x, 94.248)} {(x, 94.2)} (['the', 'length', 'of', 'a', "clock's", 'minute', 'hand', 'is', 'n', 'cm', ',', 'calculate', 'the', 'distance', 'traveled', 'by', 'its', 'tip', 'in', 'n', 'h', '.'], [['x', '=', '2', '*', '3.14', '*', 'N0', '*', 'N1']], ['15', '1'], [8, 19], {'1', '2', '3.14', '100', '60'}, '94.248') {'type': -1, 'text': "The length of a clock's minute hand is 15 cm, calculate the distance traveled by its tip in 1h.", 'flag': 0, 'original_text': "MATH QUESTION?grade 8?\nThe length of a clock's minute hand is 15 cm, calculate the distance traveled by its tip in 1h.\r\nI have no idea how to solve it, just know it has to do with circumference and arcs.", 'ans': '94.248', 'equations': 'unkn: m\r\nequ: m = 2*3.14*15*1', 'id': 'yahoo.answers.20090203143653AAKmmgN'}
[x + y - 96, x - y - 40]
r_ans, p_ans:  {(x, 28.0)} {(y, 28.0), (x, 68.0)} (['ashley', 'can', 'kayak', 'at', 'n', 'km/h', 'downstream', 'but', 'only', 'n', 'km/h', 'upstream', '.', 'what', 'is',

[0.3*x - 16.0]
r_ans, p_ans:  {(x, 53.333), (y, 26.667)} {(x, 53.333)} (['how', 'many', 'gallons', 'each', 'of', 'n', 'acid', 'and', 'n', 'acid', 'must', 'be', 'combined', 'to', 'get', 'n', 'gallons', 'of', 'solution', 'that', 'is', 'n', 'acid', '?'], [['N0', '*', 'x', '+', 'N1', '*', '(', 'N2', '-', 'x', ')', '=', 'N3', '*', 'N2']], ['60/100', '30/100', '80', '50/100'], [5, 8, 15, 21], {'1', '100', '2'}, '160/3; 80/3') {'type': -1, 'text': 'How many gallons each of 60% acid and 30 % acid must be combined to get 80 gallons of solution that is 50% acid?', 'flag': 0, 'original_text': 'This is a mixture problem... please help?\nHow many gallons each of 60% acid and 30 % acid must be combined to get 80 gallons of solution that is 50% acid?   I cannot come up with a good formula that will help me to solve this.  Is there one that  can hang with?  Thanks.', 'ans': '160/3; 80/3', 'equations': 'unkn: x,80-x\r\nequ: 0.60*x + 0.30*(80 - x) = 0.50*80', 'id': 'yahoo.answers.20090228175632AAysoZt'}

[x - 5.14893617021277]
r_ans, p_ans:  {(x, 5.447)} {(x, 5.149)} (['how', 'many', 'pieces', 'n', 'inches', 'long', 'can', 'be', 'cut', 'from', 'a', 'strip', 'of', 'metal', 'n', 'inches', 'long', '?'], [['x', '=', '(', 'N1', ')', '/', '(', 'N0', ')']], ['47/16', '121/8'], [3, 14], {'1', '12', '2'}, '5.447') {'type': -1, 'text': 'how many pieces 2 15/16 inches long can be cut from a strip of metal 15 1/8 inches long?', 'flag': 0, 'original_text': 'how many pieces 2 15/16 inches long can be cut from a strip of metal 15 1/8 inches long?\n', 'ans': '5.447', 'equations': 'unkn: x\r\nequ: x=(121/8)/(47/16)', 'id': 'yahoo.answers.20080923134431AAhuoWD'}
[12 - 2*z]
r_ans, p_ans:  {(x, 30.0), (y, 30.0)} {(z, 6.0)} (['if', 'n', 'x°', 'and', '(', 'n', 'x', '-', 'n', ')', '°', 'are', 'the', 'measures', 'for', 'vertical', 'angles', 'formed', 'by', '2', 'intersecting', 'lines', '?', 'what', 'is', 'the', 'measure', 'of', 'each', 'angle', '?'], [['N0', '*', 'z', '=', 'N1', '*', 'z', '-', 'N2']], ['5', '

[0.333333333333333 - x/500]
r_ans, p_ans:  {(x, 166.0)} {(x, 166.667)} (['if', 'n', 'out', 'of', 'n', 'boxes', 'of', 'cereal', 'have', 'a', 'prize', 'in', 'it', ',', 'about', 'how', 'many', 'boxes', 'out', 'of', 'n', 'will', 'have', 'a', 'prize', 'in', 'it', '?'], [['N0', '/', 'N1', '=', 'x', '/', 'N2']], ['1', '3', '500'], [1, 4, 20], {'1', '2'}, '166') {'type': -1, 'text': 'If 1 out of 3 boxes of cereal have a prize in it, about how many boxes out of 500 will have a prize in it?', 'flag': 0, 'original_text': 'If 1 out of 3 boxes of cereal have a prize in it, about how many boxes out of 500 will have a prize in it?\n', 'ans': '166', 'equations': 'unkn: x\r\nequ: 1/3=x/500', 'id': 'yahoo.answers.20081003092349AA9LmEA'}
[x + y - 1600, 0.06*x + 0.05*y - 85]
r_ans, p_ans:  {(x, 1100.0)} {(y, 1100.0), (x, 500.0)} (['sam', 'invested', '$', 'n', ',', 'part', 'at', 'n', 'and', 'the', 'rest', 'at', 'n', '.', 'the', 'money', 'earned', '$', 'n', 'in', '1', 'year', '.', 'how', 'much', 'was', 'inv

[y + z - 28, 6.4*y + 5.2*z - 167.2]
r_ans, p_ans:  {(x, 18.0), (y, 10.0)} {(z, 10.0), (y, 18.0)} (['fiona', 'is', 'buying', 'tulip', 'bulbs', ',', 'red', 'tulips', 'cost', '$', 'n', 'each', 'while', 'yellow', 'tulip', 'bulbs', 'cost', '$', 'n', '.', 'if', 'n', 'bulbs', 'cost', 'fiona', 'n', ',', 'how', 'many', 'of', 'each', 'type', 'did', 'she', 'buy', '.'], [['y', '+', 'z', '=', 'N2'], ['N0', '*', 'y', '+', 'N1', '*', 'z', '=', 'N3']], ['6.40', '5.20', '28', '167.20'], [10, 18, 21, 25], {'1', '2'}, '18; 10') {'type': -1, 'text': 'Fiona is buying tulip bulbs, Red tulips cost $6.40 each while yellow tulip bulbs cost $5.20. If 28 bulbs cost Fiona 167.20, how many of each type did she buy.', 'flag': 0, 'original_text': 'Another Linear Equations question?\n\n\n\nFiona is buying tulip bulbs, Red tulips cost $6.40 each while yellow tulip bulbs cost $5.20. If 28 bulbs cost Fiona 167.20, how many of each type did she buy.Please tell me how you worked it out. Thanks\n\n', 'ans': '18; 10', 'equa

[-10 + 15/(x + 2) + 24/x]
r_ans, p_ans:  {(y, 5.338), (x, 3.338)} {(x, -1.438), (x, 3.338)} (['a', 'girl', 'spends', '$', 'n', 'and', '$', 'n', 'in', 'buying', 'chicken', 'and', 'duck', 'respectively', '.', 'the', 'price', 'of', '1', 'kg', 'of', 'duck', 'is', '$', 'n', 'more', 'than', 'chicken', '.', 'if', 'the', 'total', 'mass', 'of', 'chicken', 'and', 'duck', 'is', 'n', 'kg', ',', 'determine', 'the', 'price', 'per', 'kg', 'for', 'chicken', 'and', 'duck', 'respectively', '.'], [['N0', '/', 'x', '+', 'N1', '/', '(', 'x', '+', 'N2', ')', '=', 'N3']], ['24', '15', '2', '10'], [4, 7, 24, 38], {'1', '1000', '100', '2'}, '3.338; 5.338') {'type': -1, 'text': 'a girl spends $24 and $15 in buying chicken and duck respectively. the price of 1kg of duck is $2 more than chicken. if the total mass of chicken and duck is 10kg, determine the price per kg for chicken and duck respectively.', 'flag': 0, 'original_text': 'help me in this add math question?\na girl spends $24 and $15 in buying chicken a

[x + y - 1000000, 0.19*x + 0.16*y - 180000.0]
r_ans, p_ans:  {(x, 333333.333)} {(y, 333333.333), (x, 666666.667)} (['wendy', 'has', 'n', 'to', 'lend', 'and', 'is', 'required', 'to', 'obtain', 'an', 'average', 'return', 'of', 'n', 'per', 'year', '.', 'if', 'she', 'can', 'lend', 'at', 'the', 'rate', 'of', 'n', 'or', 'at', 'the', 'rate', 'of', 'n', ',', 'how', 'much', 'can', 'she', 'lend', 'at', 'the', 'n', 'rate', 'and', 'still', 'meet', 'her', 'requirements', '?'], [['x', '+', 'y', '=', 'N0'], ['N2', '*', 'x', '+', 'N4', '*', 'y', '=', 'N1', '*', 'N0']], ['1000000', '18/100', '19/100', '16/100', '16/100'], [2, 14, 26, 32, 41], {'1', '12', '100', '2'}, '1000000/3') {'type': -1, 'text': 'Wendy has 1,000,000 to lend and is required to obtain an average return of 18% per year. If she can lend at the rate of 19% or at the rate of 16%, how much can she lend at the 16% rate and still meet her requirements?', 'flag': 0, 'original_text': 'Difficult word/equation problem! SOS!!!!?\nWendy has 1,00

[y + z - 43, y - 2*z + 5]
r_ans, p_ans:  {(y, 16.0), (x, 27.0)} {(y, 27.0), (z, 16.0)} (['a', 'scored', 'n', 'points', 'less', 'than', 'n', 'times', 'the', 'number', 'scored', 'by', 'b', '.', 'they', 'scored', 'a', 'total', 'of', 'n', 'points', '.', 'how', 'many', 'points', 'were', 'scored', 'by', 'each', 'player', '?'], [['y', '+', 'z', '=', 'N2'], ['y', '=', 'N1', '*', 'z', '-', 'N0']], ['5', '2', '43'], [2, 6, 19], {'1', '2'}, '27; 16') {'type': -1, 'text': 'A scored 5 points less than twice the number scored by B. They scored a total of 43 points. how many points were scored by each player?', 'flag': 0, 'original_text': 'A scored 5 pts less than twice the number scored by B. 2gether dey scored a total of 43 pts. how many pts...?\n\n\n\nA scored 5 pts less than twice the number scored by B. 2gether dey scored a total of 43 pts. how many pts were scored by each player?\n\n', 'ans': '27; 16', 'equations': 'unkn: x;y\r\nequ: x+y=43\r\nequ: x=2y-5', 'id': 'yahoo.answers.20070518184521AA

[y + z - 3000, 0.05*y + 0.08*z - 213]
r_ans, p_ans:  {(y, 2100.0), (x, 900.0)} {(y, 900.0), (z, 2100.0)} (['larry', 'has', 'an', 'annual', 'return', 'of', '$', 'n', 'from', 'from', '$', 'n', 'invested', 'at', 'simple', 'intrest', ',', 'some', 'at', 'n', 'and', 'the', 'rest', 'at', 'n', '.', 'how', 'much', 'is', 'invested', 'at', 'each', 'rate', '?'], [['y', '+', 'z', '=', 'N1'], ['N2', '*', 'y', '+', 'N3', '*', 'z', '=', 'N0']], ['213', '3000', '5/100', '8/100'], [7, 11, 19, 24], {'1', '12', '100', '2'}, '900; 2100') {'type': -1, 'text': 'Larry has an annual return of $213 from from $3000 invested at simple intrest, some at 5% and the rest at 8%. How much is invested at each rate?', 'flag': 0, 'original_text': 'Larry has an annual return of $213 from from $3000 invested at simple intrest, some at 5% and the rest at 8%.?\nHow much is invested at each rate?\r\n\r\nThanks if you can give me the anwser and the step by step instruction to show me how you did it!', 'ans': '900; 2100', 'equat

[x - 1.6001095890411]
r_ans, p_ans:  {(x, 1.601)} {(x, 1.6)} (['find', 'the', 'distance', 'that', 'the', 'earth', 'travels', 'in', '1', 'day', 'in', 'its', 'path', 'around', 'the', 'sun', '.', 'assume', 'that', 'a', 'year', 'has', 'n', 'days', 'and', 'that', 'the', 'path', 'of', 'the', 'earth', 'around', 'the', 'sun', 'is', 'a', 'circle', 'of', 'radius', 'n', 'million', 'miles', '.', 'please', 'help', 'me', 'find', 'the', 'answer', 'in', 'million', 'miles', '.'], [['x', '=', '2', '*', '3.14', '*', 'N1', '/', 'N0']], ['365', '93'], [22, 39], {'30', '1', '12', '2', '3.14', '100', '1000'}, '1.601') {'type': -1, 'text': 'Find the distance that the earth travels in 1 day in its path around the sun. Assume that a year has 365 days and that the path of the earth around the sun is a circle of radius 93 million miles. Please help me find the answer in million miles.', 'flag': 0, 'original_text': 'Trig question for you people out there.?\nFind the distance that the earth travels in 1 day in its 

[x - 0.833333333333333]
r_ans, p_ans:  {(x, 0.83)} {(x, 0.833)} (["miley's", 'n', '-', 'hour', 'dvr', 'has', 'n', 'gigabytes', '(', 'gb', ')', 'of', 'storage', 'space', '.', 'how', 'many', 'gb', 'are', 'used', 'to', 'store', '1', 'hour', 'of', 'regular', 'programming', '?', '(', 'round', 'to', 'the', 'nearest', 'hundredth', ')'], [['x', '=', 'N1', '/', 'N0']], ['300', '250'], [1, 6], {'1', '60', '100', '2'}, '0.83') {'type': -1, 'text': "Miley's 300-hour DVR has 250 gigabytes (GB) of storage space. How many GB are used to store 1 hour of regular programming? (round to the nearest hundredth)", 'flag': 0, 'original_text': "A math problem, 10 points for best answer?\nhere is the question : \r\n\r\nMiley's 300-hour DVR has 250 gigabytes (GB) of storage space. How many GB are used to store 1 hour of regular programming? (round to the nearest hundredth)", 'ans': '0.83', 'equations': 'unkn: x\r\nequ: x=250/300', 'id': 'yahoo.answers.20080213184554AA18G7m'}
[x - 4.292]
r_ans, p_ans:  {(x, 4.29

[x - 4*y + 8, 2*x + 2*y - 78]
r_ans, p_ans:  {(x, 9.4), (y, 29.6)} {(x, 29.6), (y, 9.4)} (['1', 'side', 'of', 'a', 'kite', 'is', 'n', 'cm', 'less', 'than', 'n', 'times', 'the', 'length', 'of', 'another', 'side', '.', 'the', 'perimeter', 'of', 'the', 'kite', 'is', 'n', 'cm', '.', 'find', 'the', 'lengths', 'of', 'the', 'sides', 'of', 'the', 'kite', '.', 'a', '.', '9.4', 'cm', 'and', '29.6', 'cm', 'b', '.', '23.5', 'cm', 'c', '.', '23.5', 'cm', 'and', '86', 'cm', 'd', '.', '9.4', 'cm'], [['x', '=', 'N1', '*', 'y', '-', 'N0'], ['2', '*', 'y', '+', '2', '*', 'x', '=', 'N2']], ['8', '4', '78'], [6, 10, 24], {'1', '100', '2'}, '9.4; 29.6') {'type': -1, 'text': 'One side of a kite is 8 cm less than four times the length of another side. The perimeter of the kite is 78 cm. Find the lengths of the sides of the kite. \n\na. 9.4 cm and 29.6 cm\nb. 23.5 cm\nc. 23.5 cm and 86 cm\nd. 9.4 cm', 'flag': 0, 'original_text': 'One side of a kite is 8 cm less than four times the length of another side. The 

[65 - x, 40 - y, -x + y + z]
r_ans, p_ans:  {(x, 65.0), (y, 40.0), (z, 15.0)} {(x, 65.0), (z, 25.0), (y, 40.0)} (['the', 'average', '(', 'arithmetic', 'mean', ')', 'of', 'n', 'numbers', 'is', 'n', '.', 'if', 'the', 'average', 'of', 'n', 'of', 'these', 'numbers', 'is', 'n', ',', 'what', 'is', 'the', 'fifth', 'number', '?'], [['N0', '*', 'N1', '=', 'x'], ['N2', '*', 'N3', '=', 'y'], ['y', '+', 'z', '=', 'x']], ['5', '13', '4', '10'], [7, 10, 16, 21], {'1', '2'}, '65; 40; 15') {'type': -1, 'text': 'The average (arithmetic mean) of 5 numbers is 13. If the average of 4 of these numbers is 10, what is the fifth number?', 'flag': 0, 'original_text': 'The average (arithmetic mean) of 5 numbers is 13. If the average of 4 of these numbers is 10, what is the fift\n\n\n\nThe average (arithmetic mean) of 5 numbers is 13. If the average of 4 of these numbers is 10, what is the fifth number?\n\n', 'ans': '65; 40; 15', 'equations': 'unkn: x,y,z\r\nequ: 5*13=x\r\nequ: 4*10=y\r\nequ: y+z=x', 'id': 'yaho

[1.75*x + 1.75*y - 560, -2*x + 2*y - 560]
r_ans, p_ans:  {(y, 20.0), (x, 300.0)} {(y, 300.0), (x, 20.0)} (['a', 'plane', 'flies', 'n', 'miles', 'in', 'n', 'hours', 'travelling', 'with', 'a', 'tailwind', '.', 'the', 'return', 'trip', 'later', 'agaisnt', 'the', 'same', 'wind', '?', 'takes', 'the', 'plane', 'n', 'hours', '.', 'what', 'were', 'the', 'speed', 'of', 'the', 'plane', 'and', 'the', 'speed', 'of', 'the', 'wind', '?'], [['(', 'y', '+', 'x', ')', '*', 'N1', '=', 'N0'], ['(', 'y', '-', 'x', ')', '*', 'N2', '=', 'N0']], ['560', '1.75', '2'], [3, 6, 25], {'1', '60', '100', '2'}, '300; 20') {'type': -1, 'text': 'A plane flies 560 miles in 1.75 hours travelling with a tailwind. The return trip later agaisnt the same wind? Takes the plane 2 hours. What were the speed of the plane and the speed of the wind?', 'flag': 0, 'original_text': 'A plane flies 560 miles in 1.75 hours travelling with a tailwind.  The return trip later agaisnt the same wind?\ntakes the plane 2 hours.  What were the

[10*x + 5*y - 32.0, 7*x + 4*y - 23.1]
r_ans, p_ans:  {(z, 23.1), (x, 2.5), (y, 1.4)} {(x, 2.5), (y, 1.4)} (['1', 'group', 'of', 'people', 'purchased', 'n', 'hot', 'dogs', 'and', 'n', 'soft', 'drinks', 'at', 'a', 'cost', 'of', '$', 'n', '.', 'a', 'second', 'bought', 'n', 'hot', 'dogs', 'and', 'n', 'soft', 'drinks', 'at', 'a', 'cost', 'of', '$', 'n', '.', 'what', 'is', 'the', 'cost', 'of', 'a', 'single', 'hot', 'dog', '?', 'a', 'single', 'soft', 'drink', '?'], [['N0', '*', 'x', '+', 'N1', '*', 'y', '=', 'N2'], ['N3', '*', 'x', '+', 'N4', '*', 'y', '=', 'N5']], ['10', '5', '32.00', '7', '4', '23.10'], [5, 9, 17, 22, 26, 34], {'1', '60', '2'}, '2.5; 1.4; 23.1') {'type': -1, 'text': 'One group of people purchased 10 hot dogs and 5 soft drinks at a cost of $32.00. A second bought 7 hot dogs and 4 soft drinks at a cost of $23.10. What is the cost of a single hot dog? A single soft drink?', 'flag': 0, 'original_text': 'Need help on this word problem...?\nOne group of people purchased 10 hot do

[99.15*x - 6.0]
r_ans, p_ans:  {(x, 40.0)} {(x, 0.061)} (['how', 'many', 'kg', 'of', 'pure', 'alcohol', 'must', 'be', 'added', 'to', 'n', 'kg', 'of', 'a', 'n', 'alcohol', 'solution', 'to', 'make', 'an', 'n', 'alcohol', 'solution', '?'], [['N0', '*', 'N1', '+', 'x', '*', '100', '=', '(', 'N0', '+', 'x', ')', '*', 'N2']], ['40', '70/100', '85/100'], [10, 14, 20], {'1', '1000', '100', '2'}, '40') {'type': -1, 'text': 'How many kg of pure alcohol must be added to 40 kg of a 70%  alcohol solution to make an 85% alcohol solution?', 'flag': 0, 'original_text': 'How many kg of pure alcohol must be added to 40 kg of a 70%  alcohol solution to make an 85% alcohol solution?\n\n', 'ans': '40', 'equations': 'unkn: x\r\nequ: 40*0.7+x*100=(40+x)*0.85', 'id': 'yahoo.answers.20080528201252AAWy917'}
[165.5 - x, 170.5 - y]
r_ans, p_ans:  {(x, 168.0), (y, 170.5)} {(x, 165.5), (y, 170.5)} (['find', 'the', 'perimeter', 'of', 'a', '41', '-', 'sided', 'polygon', 'with', 'n', 'sides', 'measuring', '(', 'n', '±

[0.375 - x]
r_ans, p_ans:  {(y, 0.562), (x, 0.375)} {(x, 0.375)} (['n', 'of', 'n', 'is', '0.375', ',', 'but', "what's", '375', 'in', 'a', 'fraction', 'form', '?'], [['N0', '*', 'N1', '=', 'x']], ['1/2', '3/4'], [0, 2], {'1', '2'}, '0.375; 0.5625') {'type': -1, 'text': "1/2 of 3/4 is 0.375, but what's 375 in a fraction form?", 'flag': 0, 'original_text': 'whats 1/2 of 3/4 in a fraction form?\n\n\n\ni know that 1/2 of 3/4 is .375, but whats .375 in a fraction form?\n\n', 'ans': '0.375; 0.5625', 'equations': 'unkn: x\r\nequ: 1/2*3/4=x', 'id': 'yahoo.answers.20070604145629AAdlSv1'}
[108.9 - 1.089*x]
r_ans, p_ans:  {(x, 100.0), (y, 125.0)} {(x, 100.0)} (['number', 'is', '$', 'n', ',', 'it', 'increase', 'n', ',', 'how', 'do', 'i', 'get', 'back', 'to', 'the', 'original', 'number', '?'], [['N0', '=', '(', '1', '+', 'N1', ')', '*', 'x']], ['108.90', '8.9/100'], [3, 7], {'1', '100', '2'}, '100; 125') {'type': -1, 'text': 'Number is $108.90, it increase 8.9%, how do I get back to the original num

[-2*x - 2*y + 244, -x + y - 72]
r_ans, p_ans:  {(x, 97.0), (y, 25.0)} {(x, 25.0), (y, 97.0)} (['the', 'perimeter', 'of', 'a', 'rectangle', 'is', 'n', 'inches.the', 'length', 'exceeds', 'the', 'width', 'by', 'n', 'inches', '.', 'what', 'is', 'the', 'length', 'and', 'the', 'width', '?'], [['N0', '=', '2', '*', '(', 'y', '+', 'x', ')'], ['y', '=', 'x', '+', 'N1']], ['244', '72'], [6, 13], {'1', '12', '2'}, '97; 25') {'type': -1, 'text': 'The perimeter of a rectangle is 244 inches.The length exceeds the width by 72 inches. What is the length and the width?', 'flag': 0, 'original_text': 'The perimeter of a rectangle is 244 inches.?\n\n\n\nBEST ANSWER\n\n\n\nThe perimeter of a rectangle is 244 inches.The length exceeds the width by 72 inches. What is the length and the width?\n\n', 'ans': '97; 25', 'equations': 'unkn: y,x\r\nequ: 244=2*(x+y)\r\nequ: x=y+72', 'id': 'yahoo.answers.20091125165810AAYiMI4'}
[1.33333333333333 - 24/x]
r_ans, p_ans:  {(x, 50.0)} {(x, 18.0)} (['the', 'frequency', 'of

[x + 1.33333333333333]
r_ans, p_ans:  {(x, -1.33)} {(x, -1.333)} (['what', 'would', 'the', 'slope', 'of', 'a', 'line', 'that', 'passes', 'through', 'the', 'points', '(', 'n', ',', 'n', ')', 'and', '(', 'n', ',', 'n', ')', 'be', '?'], [['x', '=', '(', 'N1', '-', 'N3', ')', '/', '(', 'N0', '-', '(', 'N2', ')', ')']], ['2', '-3', '-4', '5'], [13, 15, 19, 21], {'1', '2'}, '-1.33') {'type': -1, 'text': 'What would the slope of a line that passes through the points (2, -3) and (-4, 5) be?', 'flag': 0, 'original_text': "What would the slope of a line that passes through the points (2,-3) and (-4,5) be?\n\n\n\ni dont know why I can't do this\n\n\n\nCan you explain the steps?\n\n", 'ans': '-1.33', 'equations': 'unkn: z\r\nequ: z=(-3-5)/(2-(-4))', 'id': 'yahoo.answers.20080923114534AAP81m8'}
[0.03*x + 0.01*y - 0.6, x + y - 60]
r_ans, p_ans:  {(x, 20.0), (y, 40.0)} {(y, 60.0), (x, 0.0)} (["ziggy's", 'famous', 'yogurt', 'blends', 'regular', 'yogurt', 'that', 'is', 'n', 'fat', 'with', 'its', 'no', 

[13.5 - x]
r_ans, p_ans:  {(x, 1.5)} {(x, 13.5)} (["what's", 'n', 'off', 'of', '$', 'n', '?'], [['(', '1', '-', 'N0', ')', '*', 'N1', '=', 'x']], ['10/100', '15'], [1, 5], {'1', '100', '2'}, '1.5') {'type': -1, 'text': "What's 10% off of $15?", 'flag': 0, 'original_text': "What's 10% off of $15?\n\n\n\nhave a coupon for 10%, want to use it on an item that costs $15\n\n", 'ans': '1.5', 'equations': 'unkn: x\r\nequ: (1-0.1)*15=x', 'id': 'yahoo.answers.20110411120807AA4K5gT'}
[-3*x + 3*y - 48, 2*x + 2*y - 48]
r_ans, p_ans:  {(y, 4.0), (x, 16.0)} {(y, 20.0), (x, 4.0)} (['chris', 'is', 'travelling', 'by', 'boat', 'up', 'the', 'river', 'and', 'it', 'took', 'him', 'n', 'h', 'to', 'cover', 'n', 'km', 'against', 'the', 'current', '.', 'it', 'then', 'took', 'him', 'n', 'h', 'for', 'the', 'return', 'trip', 'with', 'the', 'current', '.', 'what', 'is', 'the', 'speed', 'of', 'the', 'boat', 'in', 'still', 'water', 'and', 'the', 'speed', 'of', 'the', 'current', '?'], [['(', 'y', '-', 'x', ')', '*', 'N

[1.28571428571429 - x, 4*x - y, 6*x - z]
r_ans, p_ans:  {(x, 1.286), (y, 5.16), (z, 7.74)} {(x, 1.286), (y, 5.143), (z, 7.714)} (['there', 'are', '3', 'dolls', '.', 'the', 'heights', 'of', 'the', 'dolls', 'are', 'in', 'the', 'ration', 'n', ':', 'n', ':', 'n', '.', 'the', 'tallest', 'doll', 'is', 'n', 'cm', '.', 'what', 'are', 'the', 'heights', 'of', 'the', 'other', '2', 'dolls', '?'], [['N3', '/', 'N2', '=', 'x'], ['N0', '*', 'x', '=', 'y'], ['N1', '*', 'x', '=', 'z']], ['4', '6', '7', '9'], [14, 16, 18, 24], {'1', '100', '2'}, '1.28571428571; 5.16; 7.74') {'type': -1, 'text': 'there are 3 dolls. The heights of the dolls are in the ration 4 : 6 : 7. The tallest doll is 9cm. What are the heights of the other two dolls?', 'flag': 0, 'original_text': 'Maths problem?\nThe problem is there are 3 dolls. The heights of the dolls are in the ration 4:6:7. The tallest doll is 9cm. What are the heights of the other two dolls? (PLEASE SHOW WORKING).\r\nthanks for your help.', 'ans': '1.28571428571

[0.45*x + 0.8*y - 36.0, x + y - 60]
r_ans, p_ans:  {(y, -12.0), (x, 34.286)} {(x, 34.286), (y, 25.714)} (['a', 'chemist', 'has', 'a', 'n', 'mixture', 'of', 'hydrochloric', 'acid', 'and', 'an', 'n', 'mixture', 'of', 'the', 'same', 'acid', '.', 'how', 'much', 'of', 'each', 'should', 'be', 'mixed', 'together', 'to', 'get', 'n', 'ml', 'of', 'a', 'n', 'solution', 'of', 'hydrochloric', 'acid', '?'], [['N0', '*', 'x', '+', 'N1', '*', 'y', '=', 'N3', '*', 'N2'], ['x', '+', 'y', '=', 'N2']], ['45/100', '80/100', '60', '60/100'], [4, 11, 28, 32], {'1', '1000', '100', '2'}, '34.2857142857; -12') {'type': -1, 'text': 'A chemist has a 45% mixture of hydrochloric acid  and an 80% mixture of the same acid. How much of each should be mixed together to get 60ml of a 60% solution of hydrochloric acid?', 'flag': 0, 'original_text': 'Word problems...mixture problems..help!! Whats the easiest method to use to get 60ml of a 60% solution?\n\n\n\nA chemist has a 45% mixture of hydrocholic acid  and an 80% mix

[50 - x]
r_ans, p_ans:  {(x, 10.0), (y, 50.0)} {(x, 50.0)} (['what', 'is', 'the', 'answer', 'to', 'the', 'equation', 'n', 'times', 'n', '?'], [['N0', '*', 'N1', '=', 'x']], ['25', '2'], [7, 9], {'1', '2'}, '10; 50') {'type': -1, 'text': 'What is the answer to the equation 25 times 2 ?', 'flag': 0, 'original_text': "What is the answer to the equation 25 times 2 ?\nok so I'm really bad at math and my brother said he would give me 20 bucks if I could answer the equation 25 times 2. Can anyone help me?", 'ans': '10; 50', 'equations': 'unkn: x\r\nequ: 25*2=x', 'id': 'yahoo.answers.20081108231811AAU8eXE'}
[2*x - 2*y + 780, 1.5*x + 1.5*y - 780]
r_ans, p_ans:  {(x, 455.0), (y, 65.0)} {(x, 65.0), (y, 455.0)} (['dave', 'is', 'flying', 'from', 'philadelphia', 'to', 'chicago', '.', 'on', 'the', 'flight', ',', 'flying', 'against', 'the', 'wind', ',', 'the', 'n', 'mile', 'trip', 'takes', 'n', 'hours', '.', 'the', 'return', 'trip', ',', 'with', 'the', 'wind', 'takes', 'n', 'hours', '.', 'find', 'the'

[x - 10*y, -2*x - 2*y + 154]
r_ans, p_ans:  {(y, 70.0), (x, 7.0)} {(x, 70.0), (y, 7.0)} (['the', 'perimeter', 'of', 'a', 'rectangle', 'is', 'n', 'feet', 'the', 'ratio', 'of', 'the', 'length', 'to', 'the', 'width', 'is', 'n', ':', '1', '.', 'find', 'the', 'length', 'and', 'width', '?'], [['x', '=', 'N1', '*', 'y'], ['N0', '=', '2', '*', 'y', '+', '2', '*', 'x']], ['154', '10'], [6, 17], {'1', '2'}, '7; 70') {'type': -1, 'text': 'The perimeter of a rectangle is 154 feet the ratio of the length to the width is 10 : 1. Find the length and width?', 'flag': 0, 'original_text': 'the perimeter of a rectangle is 154 feet the ratio of the lengthto the width is 10:1 find the length and width?\n\n\n\nhelp please\n\n', 'ans': '7; 70', 'equations': 'unkn: y,x\r\nequ: y=10*x\r\nequ: 154=2*x+2*y', 'id': 'yahoo.answers.20110113133927AAdsIhI'}
[3*x + 3*y - 600, -4*x + 4*y - 600]
r_ans, p_ans:  {(x, 175.0), (y, 25.0)} {(x, 25.0), (y, 175.0)} (['with', 'a', 'tail', 'wind', ',', 'a', 'small', 'piper', 'air

[5*x - 4*y, x + y - 160000]
r_ans, p_ans:  {(y, 71111.111), (x, 88888.889)} {(x, 71111.111), (y, 88888.889)} (['the', 'profits', 'for', 'a', 'partnership', 'in', 'the', 'previous', 'year', 'were', '$', 'n', '.', 'there', 'were', '2', 'partners', 'who', 'shared', 'the', 'profits', 'in', 'the', 'ratio', 'n', ':', 'n', '.', 'find', 'the', 'amount', 'of', 'profit', 'each', 'received', '.'], [['N1', '*', 'x', '=', 'N2', '*', 'y'], ['y', '+', 'x', '=', 'N0']], ['160000', '5', '4'], [11, 24, 26], {'1', '12', '2'}, '88888.8888889; 71111.1111111') {'type': -1, 'text': 'The profits for a partnership in the previous year were $160,000. There were two partners who shared the profits in the ratio 5 : 4. Find the amount of profit each received.', 'flag': 0, 'original_text': 'help with ratios (10 points to best answer)?\nThe profits for a partnership in the previous year were $160, 000. There were two partners who shared the profits in the ratio 5:4. Find the amount of profit each recieved', 'ans': '

In [500]:
set([(1,2), (2,3), (3,4)])

{(1, 2), (2, 3), (3, 4)}

In [576]:
print(cnt)

3942


In [30]:
pairs[idx]

NameError: name 'idx' is not defined

### 预处理后的数据转化为UTIE公式数据

In [9]:
extend_words = [{'word':'x', 'id': 'unk1'},
                 {'word':'y', 'id': 'unk2'},
                 {'word':'z', 'id': 'unk3'}]
extend_ents = [{'id': 'x', 'tokens': 'unk1', 'type': 'unk'},
                 {'id': 'y', 'tokens': 'unk2', 'type': 'unk'},
                 {'id': 'z', 'tokens': 'unk3', 'type': 'unk'}]
implied_words = [{'word': '1', 'id': 'sp1'},
                 {'word': '4', 'id': 'sp4'},
                 {'word': '5', 'id': 'sp5'},
                 {'word': '0', 'id': 'sp0'},
                 {'word': '6', 'id': 'sp6'},
                 {'word': '60', 'id': 'sp60'},
                 {'word': '-1', 'id': 'sp-1'},
                 {'word': '2', 'id': 'sp2'}, 
                 {'word': '100', 'id': 'sp100'},
                 {'word': '360', 'id': 'sp360'},
                 {'word': '1000', 'id': 'sp1000'},
                 {'word': '10', 'id': 'sp10'},
                 {'word': '3', 'id': 'sp3'},
                 {'word': '30', 'id': 'sp30'},
                 {'word': '12', 'id': 'sp12'},
                 {'word': '0.01', 'id': 'sp0.01'},
                 {'word': '90', 'id': 'sp90'},
                 {'word': '180', 'id': 'sp180'},
                 {'word': 'pi', 'id': 'sp_pi'}]
implied_ents = [{'id': '1', 'tokens': 'sp1', 'type': 'num'},
                {'id': '4', 'tokens': 'sp4', 'type': 'num'},
                {'id': '5', 'tokens': 'sp5', 'type': 'num'},
                {'id': '6', 'tokens': 'sp6', 'type': 'num'},
                {'id': '60', 'tokens': 'sp60', 'type': 'num'},
                {'id': '0', 'tokens': 'sp0', 'type': 'num'}, 
                {'id': '-1', 'tokens': 'sp-1', 'type': 'num'}, 
                {'id': '2', 'tokens': 'sp2', 'type': 'num'}, 
                {'id': '100', 'tokens': 'sp100', 'type': 'num'},
                {'id': '360', 'tokens': 'sp360', 'type': 'num'},
                {'id': '1000', 'tokens': 'sp1000', 'type': 'num'},
                {'id': '90', 'tokens': 'sp90', 'type': 'num'},
                {'id': '180', 'tokens': 'sp180', 'type': 'num'},
                {'id': '10', 'tokens': 'sp10', 'type': 'num'},
                {'id': '12', 'tokens': 'sp12', 'type': 'num'},
                {'id': '3', 'tokens': 'sp3', 'type': 'num'},
                {'id': '30', 'tokens': 'sp30', 'type': 'num'},
                {'id': '0.01', 'tokens': 'sp0.01', 'type': 'num'},
                {'id': 'pi', 'tokens': 'sp_pi', 'type': 'num'}]
imp_word_dict = {w['word']:w for w in implied_words}
imp_ent_dict = {w['id']: w for w in implied_ents}

In [10]:
new_data = []
from collections import OrderedDict

for I, pid in enumerate(pairs):
    x = pairs[pid]
    infix_expressions = x[1]
    relations = []
    imply_nums  = x[-2]
    imply_ns = []
    for n in imply_nums:
        if n == '3.14':
            imply_ns.append('pi')
        else:
            imply_ns.append(n)
    imply_words = [imp_word_dict[i_num] for i_num in imply_ns]
    imply_ents = [imp_ent_dict[i_num] for i_num in imply_ns]
    extend_words = [{'word':'x', 'id': 'unk1'},
                 {'word':'y', 'id': 'unk2'},
                 {'word':'z', 'id': 'unk3'}]
    extend_ents = [{'id': 'x', 'tokens': 'unk1', 'type': 'unk'},
                 {'id': 'y', 'tokens': 'unk2', 'type': 'unk'},
                 {'id': 'z', 'tokens': 'unk3', 'type': 'unk'}]
    extend_words.extend(imply_words)
    extend_ents.extend(imply_ents)
    for i, infix_expression in enumerate(infix_expressions):
        infix_exp = []
        for exp in infix_expression:
            if exp == '3.14':
                infix_exp.append('pi')
                print(infix_expression)
            else:
                infix_exp.append(exp)
        infix_expression = infix_exp
        try:
            relations.extend(infix_to_relation_with_equal(i, len(relations), infix_expression))
        except:
            print('failed to relations: ', infix_expression)
    w_dict = {'”': '"', '“': '"', u'”': '"', u'“': '"', '\u3000': ' ', '\uff1f': ' '}
    sample = {
        'info': {'sid': str(I), 'q': x},
        'words': [{'word': w_dict.get(w, w), 'id': str(i)} for i, w in enumerate(x[0])] + extend_words,
        'entities': [{'id': 'N{}'.format(i), 'tokens': str(v), 'type': 'num'} for i, v in enumerate(x[3])] + extend_ents,
        'relations': relations
    }
    add_significant_number(sample)
    try:
        sample = reid(sample, unordered_types=['+', '*', '='], readable=False)
        sample['relations'] = list(OrderedDict((r['id'], r) for r in sample['relations']).values())
#     validation_check(sample)
        new_data.append(sample)
    except:
        print('reid falied: ',sample)
#     print(sample)

failed to relations:  ['x', '=', 'N2', 'N3', '*', 'y']
['N1', '*', 'x', '=', '2', '*', '3.14', '*', 'N0']
reid falied:  {'info': {'sid': '549', 'q': (['you', 'inherit', '$', 'n', 'with', 'the', 'stipulation', 'that', 'for', 'the', 'first', 'year', 'the', 'money', 'must', 'be', 'invested', 'in', '2', 'stocks', 'paying', '6', 'percent', 'and', '11', 'percent', 'annual', 'interest', ',', 'respectively', '.', 'how', 'much', 'should', 'be', 'invested', 'at', 'each', 'rate', 'if', 'the', 'total', 'interest', 'earned', 'for', 'the', 'year', 'is', 'to', 'be', '$', 'n', '?'], [['x', '+', 'y', '=', 'N0'], ['0.06', '*', 'x', '+', '0.11', '*', 'y', '=', 'N1']], ['10000', '800'], [3, 51], ['12', '100', '1', '2'], '{6000; 4000}')}, 'words': [{'word': 'you', 'id': '0'}, {'word': 'inherit', 'id': '1'}, {'word': '$', 'id': '2'}, {'word': 'n', 'id': '3'}, {'word': 'with', 'id': '4'}, {'word': 'the', 'id': '5'}, {'word': 'stipulation', 'id': '6'}, {'word': 'that', 'id': '7'}, {'word': 'for', 'id': '8'}, 

['N0', '=', '2', '*', '3.14', '*', 'x']
reid falied:  {'info': {'sid': '4034', 'q': (['how', 'many', 'oz', 'of', 'pure', 'gold', 'must', 'be', 'added', 'to', 'n', 'oz', 'of', 'something', 'that', 'is', 'n', 'gold', 'so', 'that', 'the', 'mixture', 'is', 'n', 'gold', '?'], [['(', 'N0', '*', 'N1', ')', '+', '(', 'x', '*', '1.00', ')', '=', '(', 'N0', '+', 'x', ')', '*', 'N2']], ['5', '1/100', '6/100'], [10, 16, 23], ['100', '1', '2'], '0.265957446809')}, 'words': [{'word': 'how', 'id': '0'}, {'word': 'many', 'id': '1'}, {'word': 'oz', 'id': '2'}, {'word': 'of', 'id': '3'}, {'word': 'pure', 'id': '4'}, {'word': 'gold', 'id': '5'}, {'word': 'must', 'id': '6'}, {'word': 'be', 'id': '7'}, {'word': 'added', 'id': '8'}, {'word': 'to', 'id': '9'}, {'word': 'n', 'id': '10'}, {'word': 'oz', 'id': '11'}, {'word': 'of', 'id': '12'}, {'word': 'something', 'id': '13'}, {'word': 'that', 'id': '14'}, {'word': 'is', 'id': '15'}, {'word': 'n', 'id': '16'}, {'word': 'gold', 'id': '17'}, {'word': 'so', 'id'

In [32]:
extend_ents

[{'id': 'x', 'tokens': 'unk1', 'type': 'unk'},
 {'id': 'y', 'tokens': 'unk2', 'type': 'unk'},
 {'id': 'z', 'tokens': 'unk3', 'type': 'unk'},
 {'id': '1000', 'tokens': 'sp1000', 'type': 'num'},
 {'id': '1', 'tokens': 'sp1', 'type': 'num'},
 {'id': '2', 'tokens': 'sp2', 'type': 'num'},
 {'id': '100', 'tokens': 'sp100', 'type': 'num'}]

In [11]:
len(new_data)

4236

#### 4406个数据中
 - 转化到UTIE数据成功的有4395个
 - reid失败11个，其中failed to relation占5个

In [624]:
pairs['yahoo.answers.20090831201107AAxWXuc']

(['how', 'do', 'you', 'divide', 'n', 'by', 'n', '?'],
 [['N0', '/', 'N1', '=', 'x']],
 ['7/8', '14'],
 [4, 6],
 '0.0625')

In [626]:
from word_problem.dolphin.text_num import Text2Int


In [627]:
s = 'twenty - eight'
print(Text2Int.text2int(s))

28


In [41]:
with open('/data/hongfeng/word_problem/dolphin_t6.json', 'w') as f:
    f.write(json.dumps(new_data))

In [588]:
import random
random.shuffle(new_data)

In [ ]:
new_data[0]

In [634]:
with open('/data/hongfeng/word_problem/dolphin_t6_train.json', 'w') as f:
    f.write(json.dumps(new_data[:-len(new_data)//5]))

In [635]:
with open('/data/hongfeng/word_problem/dolphin_t6_valid.json', 'w') as f:
    f.write(json.dumps(new_data[-len(new_data)//5:]))

In [12]:
from utie.data.training_dbs import Database

In [13]:
bert_data = Database('/data/hongfeng/word_problem/dolphin_t6_bert.db/')

In [ ]:
for i, b_data in enumerate():
    print(b_data)

In [40]:
len(bert_data)

4233

In [485]:
len(bert_data)

726

### 计算模型结果正确率

In [2]:
import json

In [184]:
results = json.load(open('/data/hongfeng/word_problem/test_records/lstm_0112_noS-all-predict.json'))
labels = json.load(open('/data/hongfeng/word_problem/test_records/lstm_0112_noS-label.json'))
predicts = json.load(open('/data/hongfeng/word_problem/test_records/lstm_0112_noS-predict.json'))

In [14]:
len(results)

726

In [162]:
bert_results = json.load(open('/data/hongfeng/word_problem/test_records/bert_0114_noS-all-predict.json'))

In [163]:
len(bert_results)

726

In [52]:
labels[0], predicts[0]

({'info': {'q': [["i'm",
     'sure',
     "it's",
     'a',
     'simple',
     'answer',
     'like',
     'my',
     'last',
     'question',
     'i',
     'asked',
     '-',
     'but',
     "i'm",
     'making',
     'sure',
     "i'm",
     'doing',
     'it',
     'correctly!',
     'the',
     'ratio',
     'of',
     'the',
     'number',
     'of',
     'adults',
     'to',
     'the',
     'number',
     'of',
     'students',
     'at',
     'the',
     'prom',
     'has',
     'to',
     'be',
     'n',
     ':',
     'n',
     '.',
     'last',
     'year',
     'there',
     'were',
     'n',
     'more',
     'students',
     'than',
     'adults',
     'at',
     'prom',
     '.',
     'if',
     'the',
     'school',
     'is',
     'expecting',
     'the',
     'same',
     'attendance',
     'this',
     'year',
     ',',
     'how',
     'many',
     'adults',
     'have',
     'to',
     'attend',
     'the',
     'prom',
     '?'],
    [['x', '/', '(', 'x', '+',

In [185]:
avail_results = []
for i in results:
    for rel in i['relations']:
        if rel['type'] == '=':
            avail_results.append(i)
print(len(avail_results))    

428


In [170]:
from sympy import *
x = Symbol('x')
y = Symbol('y')
z = Symbol('z')

In [166]:
avail_results[0]

{'info': {'q': [['how',
    'to',
    'do',
    '7/53',
    'by',
    'hand',
    '?',
    '(',
    'n',
    'divided',
    'by',
    'n',
    ')',
    '?'],
   [['N0', '/', 'N1', '=', 'x'],
    ['(', 'N1', ')', '/', '(', 'N1', ')', '=', 'y']],
   ['53', '7'],
   [8, 11],
   '7.57142857143; 1'],
  'sindex': 2,
  'sid': '3251'},
 'entities': [{'tokens': 'w10|n', 'type': 'num', 'id': '(num:w10|n)'},
  {'tokens': 'w13|n', 'type': 'num', 'id': '(num:w13|n)'},
  {'tokens': 'w18|1', 'type': 'num', 'id': '(num:w18|1)'},
  {'tokens': 'w19|π', 'type': 'num', 'id': '(num:w19|π)'},
  {'tokens': 'w16|x', 'type': 'unk', 'id': '(unk:w16|x)'},
  {'tokens': 'w17|y', 'type': 'unk', 'id': '(unk:w17|y)'}],
 'relations': [{'layer': 0,
   'type': '/',
   'id': '[/]((num:w10|n),(num:w13|n))',
   'operands': ['(num:w10|n)', '(num:w13|n)'],
   'prob': 0.9999488592147827},
  {'layer': 1,
   'type': '=',
   'id': '[=]((unk:w17|y),[/]((num:w10|n),(num:w13|n)))',
   'operands': ['(unk:w17|y)', '[/]((num:w10|n),(n

In [186]:
equations = []
def replace_eq(op, rel2eq):
    # 从长找到短
    for rel in rel2eq:
        if rel[0] in op:
            op = op.replace(rel[0], rel[1])
            replace_eq(op, rel2eq)
    else:
        return op
    
def replace_v(op, wid2value):
    for wid, v in wid2value.items():
        if wid in op:
            op = op.replace(wid, v)
            replace_v(op, wid2value)
    else:
        return op
eq_ans = []
for sam in avail_results:
    rel2eq = {}
    eqs = []
    _, _, values, _, ans = sam['info']['q']
    idx = 0
    wid2value = {}
    ent2v = {}
    for i, w in enumerate(sam['words']):
        if w['word'] == 'n' and not sam['words'][i+1]['word'].startswith('##'):
            wid2value[w['id']] = values[idx]
            idx += 1
        else:
            wid2value[w['id']] = w['word']
    for e in sam['entities']:
        if e['tokens'] in wid2value:
            if '/' in wid2value[e['tokens']]:
                ent2v[e['id']] = '({})'.format(wid2value[e['tokens']])
            else:
                ent2v[e['id']] = wid2value[e['tokens']]
    for rel in sam['relations']:
        if rel['type'] != '=':
            rel2eq[rel['id']] = '({}{}{})'.format(rel['operands'][0], rel['type'], rel['operands'][1])
        if rel['type'] == '=':
            eqs.append('{}-{}'.format(rel['operands'][0], rel['operands'][1]))
    rel2eq = sorted(rel2eq.items(), key=lambda x: len(x[0]), reverse=True)
    new_eqs = []
    for eq in eqs:
        eq = replace_eq(eq, rel2eq)
        eq = replace_v(eq, ent2v)
        new_eqs.append(eval(eq))
    eq_ans.append((new_eqs, ans))
    print('origin_data: ',sam['info']['q'], sam['relations'])
    print('new_eqs: ', new_eqs)

origin_data:  [['n', 'divided', 'by', 'n', 'is', '?'], [['N0', '/', '(', 'N1', ')', '=', 'x']], ['17/5', '4/3'], [0, 3], ' 2.55'] [{'layer': 0, 'type': '/', 'id': '[/]((num:w0|n),(num:w3|n))', 'operands': ['(num:w0|n)', '(num:w3|n)'], 'prob': 0.9994398951530457}, {'layer': 1, 'type': '=', 'id': '[=]((unk:w6|x),[/]((num:w0|n),(num:w3|n)))', 'operands': ['(unk:w6|x)', '[/]((num:w0|n),(num:w3|n))'], 'prob': 1.0}]
new_eqs:  [x - 2.55]
origin_data:  [['n', 'is', 'to', 'n', 'as', 'n', 'is', 'to', '?'], [['N0', '/', 'N1', '=', 'N2', '/', 'x']], ['7', '34', '5'], [0, 3, 5], '24.2857142857'] [{'layer': 0, 'type': '/', 'id': '[/]((num:w0|n),(num:w3|n))', 'operands': ['(num:w0|n)', '(num:w3|n)'], 'prob': 0.9995200634002686}, {'layer': 0, 'type': '/', 'id': '[/]((num:w5|n),(unk:w9|x))', 'operands': ['(num:w5|n)', '(unk:w9|x)'], 'prob': 0.8788811564445496}, {'layer': 1, 'type': '=', 'id': '[=]([/]((num:w0|n),(num:w3|n)),[/]((num:w5|n),(unk:w9|x)))', 'operands': ['[/]((num:w0|n),(num:w3|n))', '[/]((

origin_data:  [['what', 'is', 'the', 'value', 'of', 'n', 'divided', 'by', 'n', 'multiplied', 'by', 'n', '?'], [['N0', '/', 'N1', '=', 'x'], ['x', '*', 'N2', '=', 'y']], ['18', '9', '3'], [5, 8, 11], '2; 6'] [{'layer': 0, 'type': '/', 'id': '[/]((num:w5|n),(num:w8|n))', 'operands': ['(num:w5|n)', '(num:w8|n)'], 'prob': 0.9994617104530334}, {'layer': 1, 'type': '=', 'id': '[=]((unk:w13|x),[/]((num:w5|n),(num:w8|n)))', 'operands': ['(unk:w13|x)', '[/]((num:w5|n),(num:w8|n))'], 'prob': 0.9999996423721313}]
new_eqs:  [x - 2.0]
origin_data:  [['what', 'is', '$', 'n', 'minus', '(', '-', ')', 'n', '?'], [['x', '=', 'N0', '(', '1', '-', 'N1', ')']], ['30.00', '15/100'], [3, 8], '25.50'] [{'layer': 0, 'type': '-', 'id': '[-]((num:w3|n),(num:w8|n))', 'operands': ['(num:w3|n)', '(num:w8|n)'], 'prob': 0.9799097776412964}, {'layer': 1, 'type': '=', 'id': '[=]((unk:w10|x),[-]((num:w3|n),(num:w8|n)))', 'operands': ['(unk:w10|x)', '[-]((num:w3|n),(num:w8|n))'], 'prob': 0.9999997615814209}]
new_eqs:  [x

new_eqs:  [-x - y + 336]
origin_data:  [['what', 'percentage', 'would', 'be', 'n', 'out', 'of', 'n', '?'], [['x', '=', 'N0', '/', 'N1']], ['35', '42'], [4, 7], '0.833'] [{'layer': 0, 'type': '/', 'id': '[/]((num:w4|n),(num:w7|n))', 'operands': ['(num:w4|n)', '(num:w7|n)'], 'prob': 0.9936680197715759}, {'layer': 1, 'type': '=', 'id': '[=]((unk:w9|x),[/]((num:w4|n),(num:w7|n)))', 'operands': ['(unk:w9|x)', '[/]((num:w4|n),(num:w7|n))'], 'prob': 1.0}]
new_eqs:  [x - 0.833333333333333]
origin_data:  [['a', 'bag', 'contains', 'n', 'red', 'marbles', ',', 'n', 'blue', 'marbles', ',', 'and', 'n', 'green', 'marbles', '.', 'if', '1', 'marble', 'is', 'selected', 'at', 'random', ',', 'determine', 'the', 'probability', 'that', 'it', 'is', 'blue', '.'], [['x', '=', 'N1', '/', '(', 'N0', '+', 'N1', '+', 'N2', ')']], ['7', '3', '2'], [3, 7, 12], '1/4'] [{'layer': 0, 'type': '*', 'id': '[*]((num:w3|n),(num:w7|n))', 'operands': ['(num:w3|n)', '(num:w7|n)'], 'prob': 0.9980313181877136}, {'layer': 1, 'typ

In [188]:
from sympy import *
from word_problem.math23k.process import to_float
x = Symbol('x')
y = Symbol('y')
z = Symbol('z')
cnt = 0
for e_a in eq_ans:
    r_ans = e_a[1].strip('{} ').split(';')
    try:
        r_ans = sorted([round(to_float(a), 5) for a in r_ans])
    except:
        print('wrong format of r_ans:', r_ans )
    p_ans = solve(e_a[0],[x,y,z])
    try:
        p_ans = sorted([round(float(v), 5) for k, v in p_ans.items()])
    except:
        print('wrong format of p_ans:', p_ans )
    if len(p_ans) == len(r_ans):
        print(p_ans, r_ans)
        for r, p in zip(r_ans, p_ans):
            if r != p:
                print('not_equal:', r, p)
                break
        else:
            cnt += 1
            print(cnt)

[2.55] [2.55]
1
[24.28571] [24.28571]
2
[12.0, 16.0] [12.0, 16.0]
3
[21.0] [21.0]
4
[6000.0, 13000.0] [6000.0, 13000.0]
5
[0.76562] [36.0]
not_equal: 36.0 0.76562
[15.0, 18.0] [15.0, 18.0]
6
[4.0] [4.0]
7
[617.83333] [1031.22]
not_equal: 1031.22 617.83333
[23.0, 24.0, 25.0] [23.0, 24.0, 25.0]
8
[21.33333] [12.0]
not_equal: 12.0 21.33333
[0.512] [0.512]
9
[54.9] [54.9]
10
[10000.0] [10000.0]
11
[0.04091] [0.04091]
12
wrong format of p_ans: []
wrong format of p_ans: []
[0.65] [0.65]
13
wrong format of p_ans: {x: 16 - y}
[150.0, 173.0] [150.0, 173.0]
14
[4.0] [4.0]
15
wrong format of p_ans: {x: 36 - 3*y/5}
[0.015] [0.015]
16
[4.44444] [4.44444]
17
[4.75] [4.75]
18
[2.66667] [2.66667]
19
wrong format of r_ans: ['18 square meters']
wrong format of p_ans: []
[1.66667] [1.66667]
20
[14.0, 64.0] [14.0, 64.0]
21
wrong format of p_ans: []
[16.0] [16.0]
22
wrong format of p_ans: {x: 54000 - y}
wrong format of p_ans: {x: 47.444921316166 - 1.5722460658083*y}
[0.06667] [0.33333]
not_equal: 0.33333 0

[16.0] [16.0]
163
[52.8] [52.8]
164
[0.09124] [0.091]
not_equal: 0.091 0.09124
[-1.93333] [1.5]
not_equal: 1.5 -1.93333
[5.0] [5.0]
165
wrong format of p_ans: {x: 40 - y}
wrong format of p_ans: []
[30.0] [62.0]
not_equal: 62.0 30.0
[20.0, 25.0] [20.0, 25.0]
166
wrong format of p_ans: []
[3.0] [3.0]
167
[306.0] [306.0]
168
[0.8] [0.8]
169
[1.16667] [1.16667]
170
[-186.66667] [47.0]
not_equal: 47.0 -186.66667
[3.21429] [11.25]
not_equal: 11.25 3.21429
[5.0] [3.0]
not_equal: 3.0 5.0
[75.0] [75.0]
171
[0.98258] [0.01773]
not_equal: 0.01773 0.98258
[10.0] [0.064]
not_equal: 0.064 10.0
[0.77778] [0.77778]
172
[0.6] [0.6]
173
wrong format of p_ans: {x: 33/10 - y/10}
[0.6] [0.6]
174
wrong format of p_ans: {x: 132 - y}
wrong format of p_ans: {x: -y - z + 2.48}
[24190.0] [0.57561]
not_equal: 0.57561 24190.0
[45.0] [45.0]
175
wrong format of p_ans: {x: 66 - y}
[140.0] [140.0]
176
[25.0, 37.0] [25.0, 37.0]
177
[14.58333] [14.58333]
178
wrong format of p_ans: {x: 90 - 3*y/2}
[45.0] [45.0]
179
[91.0

In [452]:
eval('x-126*4/9')

x - 56.0

In [456]:
cnt

4097

In [401]:
solve([x - y - 13, x * y - 950])
# data_dict['yahoo.answers.20080906193908aac44in']

[{x: -25, y: -38}, {x: 38, y: 25}]

In [400]:
solve([x - y - 13, x + y - 3])

{x: 8, y: -5}

In [391]:
ans_ = []
for i in ans:
    ans_.append({(c, round(float(i[c]),3)) for c in i})
ans_

[{(x, -25.0), (y, -38.0)}, {(x, 38.0), (y, 25.0)}]

In [448]:
s = 'Julia'
re.sub('km\^2|/?[a-zA-Z]|\s', '', s)

''

In [399]:
solve([z+4-y, z/y-2/3, x-z/y])

[{x: 0.666666666666667, y: 12.0000000000000, z: 8.00000000000000}]

In [336]:
import regex as re
a = '5/3 kg/m'
a = re.sub('/?[a-z]', '', a)
a

'5/3 '

In [373]:
data_dict['yahoo.answers.20070520185706aacdvlu']

{'type': 0,
 'text': 'The denominator of a fraction exceeds the numerator by 4. When the fraction is reduced to lowest terms it is 2/3. What is the fraction?',
 'sources': ['https://answers.yahoo.com/question/index?qid=20070520185706AACDvLu'],
 'flag': 0,
 'original_text': '',
 'ans': '8/12',
 'equations': 'unkn: f\r\nequ: d = n + 4\r\nequ: n/d = 2/3\r\nequ: f = n/d',
 'id': 'yahoo.answers.20070520185706aacdvlu'}

In [325]:
data_dict['yahoo.answers.20070516170746aa02o0j']

{'type': 0,
 'text': 'Find two numbers whose sum is 32 and such that one is 3 times as large as the other?',
 'sources': ['https://answers.yahoo.com/question/index?qid=20070516170746AA02o0j'],
 'flag': 0,
 'original_text': '',
 'ans': '24; 8',
 'equations': 'unkn: x, y\r\nequ: x + y = 32\r\nequ: x  = 3*y',
 'id': 'yahoo.answers.20070516170746aa02o0j'}

In [370]:
data_dict['yahoo.answers.20070823215055aacjodn']

{'type': 0,
 'text': 'x + 20 = y + 80. What is the value of x-y.',
 'sources': ['https://answers.yahoo.com/question/index?qid=20070823215055AAcjOdN'],
 'flag': 0,
 'original_text': '',
 'ans': '60',
 'equations': 'unkn: n\r\nequ: x + 20 = y + 80\r\nequ: n = x - y',
 'id': 'yahoo.answers.20070823215055aacjodn'}

In [377]:
data_dict['yahoo.answers.20080419220125aasopis']


{'type': 0,
 'text': 'Five times a number is equal to another number. Their difference is 52. Find the smaller number.',
 'sources': ['https://answers.yahoo.com/question/index?qid=20080419220125AASopis'],
 'flag': 0,
 'original_text': '',
 'ans': '13 | -65',
 'equations': 'unkn: y\r\nequ: 5*x = y\r\nequ: x - y = 52',
 'id': 'yahoo.answers.20080419220125aasopis'}

In [291]:
s = '2x+3y-5.2(1-x)-(1/2)x+(10+99)(99-10+1)+(1/2)5'

In [449]:
solve([x + 1/x - 13/6])

[{x: 0.666666666666667}, {x: 1.50000000000000}]

In [439]:
cnt

4059

In [287]:
import regex as re
comma_in_num_re = re.compile(r'(\d+),(\d+)(\.\d+)?')

In [301]:
omit1 = re.compile('(\d+)([a-z])')
omit2 = re.compile('(\d+|[a-z])(\()')
omit3 = re.compile('(\))(\d+|[a-z])') 
omit4 = re.compile('(\))(\()')                                    

In [302]:
def add_mul_in_eq(text):
    # 把 12,345 变成 12345
    def comma_f(m):
        groups = m.groups()
        return '{}*{}'.format(groups[0], groups[1])
    text = omit1.sub(repl=comma_f, string=text)
    text = omit2.sub(repl=comma_f, string=text)
    text = omit3.sub(repl=comma_f, string=text)
    text = omit4.sub(repl=comma_f, string=text)
    return text

In [303]:
add_mul_in_eq(s)

('2', 'x')
('3', 'y')
('2', '(')
(')', 'x')
(')', '5')
(')', '(')


'2*x+3*y-5.2*(1-x)-(1/2)*x+(10+99)*(99-10+1)+(1/2)*5'

In [264]:
data[0]

{'type': 0,
 'text': 'one number is 4 less than another. the difference  of twice the smaller and 5 times the larger number is -11. What are the numbers?',
 'sources': ['https://answers.yahoo.com/question/index?qid=20080421132620AAXFXiG'],
 'flag': 0,
 'original_text': '',
 'ans': '-3; 1',
 'equations': 'unkn: x, y\r\nequ: x = y - 4\r\nequ: 2*x - 5*y = -11',
 'id': 'yahoo.answers.20080421132620aaxfxig'}

In [275]:
data_dict = {}
for d in data:
    data_dict[d['id']] = d

In [320]:
data_dict['yahoo.answers.20070823215055aacjodn']

{'type': 0,
 'text': 'x + 20 = y + 80. What is the value of x-y.',
 'sources': ['https://answers.yahoo.com/question/index?qid=20070823215055AAcjOdN'],
 'flag': 0,
 'original_text': '',
 'ans': '60',
 'equations': 'unkn: n\r\nequ: x + 20 = y + 80\r\nequ: n = x - y',
 'id': 'yahoo.answers.20070823215055aacjodn'}

In [266]:
len(data_dict)

6762

In [271]:
for k,d in data_dict.items():
    print(k, d)
    break

yahoo.answers.20080421132620aaxfxig {'type': 0, 'text': 'one number is 4 less than another. the difference  of twice the smaller and 5 times the larger number is -11. What are the numbers?', 'sources': ['https://answers.yahoo.com/question/index?qid=20080421132620AAXFXiG'], 'flag': 0, 'original_text': '', 'ans': '-3; 1', 'equations': 'unkn: x, y\r\nequ: x = y - 4\r\nequ: 2*x - 5*y = -11', 'id': 'yahoo.answers.20080421132620aaxfxig'}


In [80]:
for rel in rel2eq:
    print(rel)

('[+]([*]((num:w14|n),(unk:w40|x)),[*]((num:w20|n),(unk:w41|y)))', '([*]((num:w14|n),(unk:w40|x))+[*]((num:w20|n),(unk:w41|y)))')
('[*]((num:w14|n),(unk:w40|x))', '((num:w14|n)*(unk:w40|x))')
('[*]((num:w20|n),(unk:w41|y))', '((num:w20|n)*(unk:w41|y))')
('[+]((unk:w40|x),(unk:w41|y))', '((unk:w40|x)+(unk:w41|y))')


In [281]:
solve([5*x-y,x - y-52], [x,y])

{x: -13, y: -65}

In [277]:
solve([11/100*12000+14/100*x-(12/100*(12000+x))], [x])

{x: 6000.00000000000}

In [278]:
eval('x-((45/100)*20)')

x - 9.0

In [224]:
to_float(' 3/2')

1.5

In [121]:
2.55000 == 2.55

True

In [124]:
54.9 == 54.90000

True

In [352]:
data_dict['yahoo.answers.20071118151421aa7zvpx']

{'type': 0,
 'text': 'What fraction when added to its reciprocal is equal to 13/6?',
 'sources': ['https://answers.yahoo.com/question/index?qid=20071118151421AA7ZVPx'],
 'flag': 0,
 'original_text': '',
 'ans': '3/2 or 2/3',
 'equations': 'unkn: f\r\nequ: f + 1/f = 13/6',
 'id': 'yahoo.answers.20071118151421aa7zvpx'}